# Meeting Notes and Things To Do
Provided in reverse chronological order

## 20180917 
1.  Change Pedigree browser button label to say "Trim pedigree based on focal animals". Done 20180917
2.  Give user the option to save a skeleton configuration file to their home directory if they do not have a configuration file.
3.  Incorporate the error detection functions into the shiny application so that if a defective pedigree file is selected by the user a new tab is created with a description of the errors detected.
    a. The user should be moved to the error report tab.
    b. The error report tab should be placed immediately after the input tab.
4.  Beth is to try the current (0.3.31) or later version of nprcmanager with the pedigree file that has been failing. 
    a. If that file still causes an error, Beth is to send it to Mark.
    b. Mark will change the code to better handle the issues observed, update the package, and have Beth test again.
    
## 20180820

Meeting was canceled due to Amanda's family emergency.

Topics I wanted to bring up.

1. Check on ability of Beth and Amanda to install package.
2. Demonstrate use of __qcStudbook__ in interactive mode.
3. See if I should follow up with users who have had trouble loading a file.
4. Demonstration to genetics and genomics group
5. Breeding group formation. Workflow and goals
6. Genetic value analyses. Workflow and goals

## 20180711 Meeting Notes

### Summary Statistics tab

1. Use default box and whisker plot with red dots for outliers.
    a. Changes made. Done 2018-07-14
    b. Hover text to box and whisker plots added. Done 2018-07-14

### Pedigree Browser tab
1. Change Update breeding colony button to "" Waiting on Amanda. Changed to "Update Focal Animals" based on instructions from Amanda. Done 2018-08-04
2. Add function to "Update Breeding Colony" button to browse for a file; add animal IDs to the window where they could be pasted. Started 2018-07-29 getting file not updating window. Done 2018-08-15
3. Change first bullet point to reflect change in function of the "Update Focal Animal" button. Done 2018-08-15
4. Move first bullet point to far right. Done 2018-08-15
5. Remove second bullet point. However, provide this information when the file has any type of error. This is complete for one error at a time. Currently the __qcStudbook__ function does not continue scanning the file when an error is found. Done 2018-07-17
6. Change "Display UIDs for partial parentage" to "Display UIDs". Consider coming up with a better name than "UID". Add hovertext to explain what this button does. Proposed explanatory text: "Unknown IDs are created by the application for all animals with only one parent. They begin with a capital __U__." Done 2018-08-15
7. Change "Trim pedigree based on specified population" to "Trim Pedigree". Add hovertext to explain what this button does. See what "Trim Pedigree" button does. Does it remove animals? Its intent is to reduce the number of animals being examined. It does reduce the number by removing all individuals not related to the focal animals. Does it change genetic analysis. It does not change the genetic analysis for the animals left. Obviously for animals removed, it does. The name of this button will likely follow from what the new name for "Update Breeding Colony" button. Done 2018-08-16
8. Add a better description for the Search field. Done 2018-08-16

### Error Handling
1. Develop a full set of error messages appropriate for all types of input files.
    a. Added improved error detection and reporting for missing required columns - Done 2018-07-17
    b. Added the ability to call qcStudbook interactively with a flag (__errors__) set to TRUE that 
    causes qcStudbook to report back a list with all errors detected within the input file.
    Not all errors can be detected in a single pass because some errors preclude checking for others.
    For example, if the birth date column has an invalid format, the parental age cannot be checked.
    If no errors are found with __errors__ set to TRUE and NULL value is returned.
    __errors__ defaults to __FALSE__.
    c. This emphasizes the need for a good set of corrupt input files for testing and unit test
    development.
2. Amanda and Beth will provide me with some files with errors.
   a. Amanda provided a file with two error types: bad birth date type - integer instead of character representation of a date and animal appearing as a sire and a dam. Done 20180720
   b. Both errors were caught, but the first error was only seen after correcting the first. It would be better to report all errors and set a flag that told the program to stop at the end of parsing the entire pedigree.
   c. Three files were added to the example files: _Example_Pedigree.csv_, 
   _Pedigree_File_Example_134M_dam_removed.csv_, and _Pedigree_File_Example_CSV.csv_ Done 20180721
   d. These files need to be documented.
3. Refactored some code within __qcStudbook__ to a separate function __unknown2NA__. Done 2018-07-18
4. Added bad date string format detection in __convertDates__, which is now __convertDate__. 
Also added unit tests for __convertDate__. Done 2018-07-21
5. Added ability to handle dates with NA to __convertDates__ Done 2018-07-22
6. Added unit test coverage of __updateProgress__ calls in _test_reportGV.R_ Done 2018-07-22
7. Corrected a bug and added corollary unit test for __rankSubjects__. Done 2018-07-29
8. Added test for missing __age__ column in _orderReport_ function. This also allows correct
ordering when the age column is missing. Done 2018-07-29
9. Added unit tests for _findLoops_ function. Done 2018-07-30
10. Added check for class on objects in _convertDates_ function, because it was failing when 
objects were already dates. It now passes those objects through untouched. Done 2018-08-04
11. Added some debug code for updating focal animals (still named breeding colony update in code) Done 2018-08-15

### Multiple tabs
1. Come up with a different word for "Breeders". What do you think about _Managed Animals_
    a. Amanda suggested _Focal Animals_ -- this is being adopted.

### Miscellaneous items
1. Added automatic generation of a web site
    a. 2018-07-17 first draft
    b. Need to first move content that has been prepared to the correct location. Done 2018-08-16
2. Changed read.csv to read.table in code to emphasize we are able to read multiple file types.
    a. 2018-07-17
3. Amended documentation for the __addGenotype__ function to indicate that it is assuming 
the _genotype_ object was opened by __checkGenotypeFile__. 2018-07-17
4. Added use of convenience functions __get_and_or_list__ and __is_valid_data_str__ from the _github/rmsharp/rmsharp/rmsutilityr_ repository. 2018-07-21
5. Added documentation for data elements __finalRpt__ and __rpt__, which are both created by the _reportGV_ 
function. 2018-07-29
6. Added brief tutorial on how to use _findLoops_ function. 2018-08-04

### Questions that came up after the meeting
1. What is the meaning of the following:
    a. Text under __Export__ button on _Pedigree Browser_ tab. 
    ``A population must be defined before proceeding to the Genetic Value Analysis''.
    b. Check box above __Export__ button on _Pedigree Browser_ tab. 
    ``Trim pedigree based on specified population''. 
    Its intent is to reduce the number of animals being examined. 
    It does reduce the number by removing all individuals not related 
    to the focal animals. Does it change genetic analysis. 
    It does not change the genetic analysis for the animals left. 
    Obviously for animals removed, it does. Done 2018-08-16
    
## 20180611 Meeting Notes

### Input tab 

1. Move Column content into description column separate with colon - Done
2. Change Column Name to Allowable Name - Done
3. Names have alphanumeric plus "_", "-", and " ". - Done
4. Remove " Other characters have not been tested." - Done
5. Change "Any animals listed as a Sire or Dam that do not have their own row or line entry as an Ego will be added." to
A new row entry will be added for any Sire or Dam that do not already have their own row as an Ego. - Done
6. Change "Parents will be checked to ensure their own Ego entry is the correct sex." to "Animals will be checked to ensure that their sex is consistent throughout the file."" - Done
9. Use Allele_1 Allele_2 - Done

### Summary Statistics tab

1. Remove background on histograms - Done
8. Add box and whisker plot to right of summary statistics histograms - Done
10. Use Glossary for definition of terms

## 20180501 Meeting Notes

1. The second bullet point under the Input File Handling section of the Input tab says "Please be aware that animals with no parents will be treated as founders in these calculations, i.e., sources of new genetic variation in the colony."

I am assuming that this is fine and stated only to keep users aware that if they have parental information it has to be entered.

2. The third bullet point under the Input File Handling section of the Input tab says "Designation of two alleles is required as currently there is no accommodation for partial genetic information for an individual."

Is this will be a problem that needs to be addressed?

3. Reminder for Mark only: The third bullet point under the Input File Handling section of the Input tab says "If the Age column is provided, the program will use the user-specified age."

Make sure this is true. I may also want to check for internal inconsistencies.

4. In the Input File Handling section of the Input tab the last bullet point immediately above the first table says "Genotype data may be supplied within the pedigree file or in a separate genotype file. Only two additional columns (first and second) are required when the genotypes are provided within the pedigree file."

This should be explained more fully and the program behavior may need to be changed. Currently the program assumes the columns first and second are there and are correctly formed. Currently there is no check to ensure that first_name and second_name data elements are consistent with first and second data elements.

5. In the Input File Handling section of the Input tab and under the first table of column names, there is a section describing what is done by the qcStudbook function to correct data and create new columns.

Should we provide a summary of changes made to the input data such as changes in column names, sex identifiers, etc., removal of duplicated rows and columns that are added?

Need name for genotype (first_name, second_name)
Need new name for breeders only file type.
Reword "Animals without birth dates are not considered."
When minimum parent age is included, animals without birth dates are not rejected.

Change update breeding colony (pedigree browser) by adding explanatory text.

## Miscellaneous Accomplishments

These should be stored elsewhere

#### 20180429
Test coverage is 86.32%.

#### 20171004
Unit test coverage is 84.79%.

#### 20170921
Unit test coverage is about 15%.

#### 20170919
We have successfully installed and used _nprcmanager_ on Microsoft (MS) 
Windows 7, MS Windows 10, and MacOS 10.12.6 running R 3.4.1.

#### 20170917
I discovered why the plot of genetic uniqueness scores was not as expected. 
It was from an analysis where 0 was the threshold setting, which I think we 
should consider removing as an option. I have set the default to 3, 
which is what Amanda Vinson's paper indicated that ONPRC typically used.

I got the logging system integrated into the package. 
I am using the package futile.logger (funny name). Note the check box at the 
bottom of the side panel on the two attached images. 
When the Debug on check box is checked (it is not checked by default), 
the application writes to a file in the users home directory named
_nprcmanager.log_. Currently I am only logging events occurring the the 
_server.R_ file, because that is where I tend to have most of my problems 
show up. I have attached an example file in which turned on logging at the 
debug level and then read in a couple of pedigree files.

I added code coverage reports to the automated build system running on 
Travis-CI.org. Currently we only have 5.68% of the code being tested with the 
unit tests I created thus far. I do not know the code well enough to know what 
percentage we will have when I feel comfortable, but I know most of the 
functions have no tests at all.

#### 20170916
Can now assign known genotypes to individuals and to 
incorporate that information into the gene dropping routine.
This has been done at the function level in the _geneDrop_ function. 
The pedigree submitted via the UI can optionally contain
genetic information. 

#### 20170915
Added a genetic uniqueness plot to the Summary Statistics tab.

#### 20170911
I have connected the Travis CI (continuous integration) tool to my 
github.com/rmsharp/nprcmanager so that when I check in code it automatically
tries to build it from scratch using packages from CRAN. Travis CI is now 
automatically building nprcmanager without errors or warnings. 
This puts us considerably closer to being able to offer the package via 
CRAN if we want.

## 20170323 Meeting notes

I met with Jack Kent, Deborah Newman, and Charles Peterson about how to use
genetic data in a gene dropping simulation.

From: R. Mark Sharp <msharp@TxBiomed.org>

Subject: Re: Genetic management of colonies

Date: March 23, 2017 at 5:01:07 PM CDT

To: Jack Kent <jkent@txbiomed.org>

Cc: Deborah Newman <dnewman@txbiomedgenetics.org>, 
Charles Peterson <charlesp@txbiomed.org>

Jack, Debbie, and Charles,

Thank you for meeting with me. I appreciate your help in clarifying the 
various stages of the issues associated with providing genetic management 
guidance via simulation with partially known MHC data. 

1. I will look more closely into what will be needed to add genetic 
information and gene frequency information into the gene dropping routines 
in the kinship2 package. The initial implementation will use only a single 
locus, which should be a sufficient for MHC data

	A. First step will be to ensure we get expected proportions of genes in 
	children of individuals when we provide genetic data to all founders using 
	a gene frequency based algorithm.

	B. Second step will be to ensure we get expected proportions of genes in 
	children when all parental generations have fully known genotypes.

	C. Third step will be to ensure we get expected proportions of genes in 
	children when parental generations have partially known genotypes and all 
	other genotypes are uninformative.

	D. Fourth step will be to ensure we get expected proportions of genes in 
	children when parental generations have partially known genotypes and 
	remaining parental genes are determined by gene frequency.

2. We will not worry about getting rid of genes in the pedigrees since colony 
managers need only not breed animals that carry unwanted alleles.

3. I will produce a small number of pedigree drawings using the data Debbie 
has provided to get feedback on preferences and additional requirements.

4. I will not deal with paternity and maternity issues within the pedigrees 
at this time.

Thanks again.

