---
title: Development Plans
author: R. Mark Sharp
date: Sept. 1, 2017
output: rmarkdown::html_vignette
vignette:
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{Development Plans}
%\usepackage[UTF-8]{inputenc}
------

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Next steps
1. Find out what is needed to assign known genotypes to individuals and to 
    incorporate that information into the gene dropping routine.
    This has been done at the function level in the _geneDrop_ function. 
    The next step is to allow the pedigree submitted via the UI to contain
    the genetic information. Completed by 20170916.
2. Find some ways to use genetic marker data in the evaluation of genetic 
   value. This needs to be discussed.
3. Develop a simplistic method of using data directly from LabKey via the 
   labkey-api-r. There is now a configuration file that can be used by 
   the application to get the primate center designation, the LabKey 
   baseUrl, a schemaName, a folderPath, and the queryName. 
   The file is structured so that more than one of each could be placed in the 
   file. However, more design work is needed before this will be of use. 
   Completed initial version 20170907.
 
4. Develop the appropriate UI - Completed 20170916.
I got the logging system integrated into the package. I am using the package futile.logger (funny name). Note the checkbox at the bottom of the side panel on the two attached images. When the Debug on checkbox is checked (it is not checked by default), the application writes to a file in the users home directory named nprcmanager.log. Currently I am only logging events occurring the the server.R file, because that is where I tend to have most of my problems show up. I have attached an example file in which turned on logging at the debug level and then read in a couple of pedigree files.

I added code coverage reports to the automated build system running on Travis-CI.org. Currently we only have 5.68% of the code being tested with the unit tests I created thus far. I do not know the code well enough to know what percentage we will have when I feel comfortable, but I know most of the functions have no tests at all.


## To do
1. It would be better to catalogue and report the changes made by qcStudbook()
   so that the user knows how the provided pedigree was changed.
2. Amanda Vinson wrote this in an e-mail on 20170901:
  
I'm very interested in hearing about your idea of adding algorithms related to 
selecting/maintaining specific alleles in a colony. 
I think your idea could be related to what I need to work on next, 
which is an algorithm for assessing the long-term impact on colony genetic 
diversity of removing specific animals/alleles.  
This is a need we have here at the ONPRC due to assigning animals to research, 
and I have been tasked with implementing the solution.


## Accomplishments thus far
#### 20170911

I have connected the Travis CI (continuous integration) tool to my 
github.com/rmsharp/nprcmanager so that when I check in code it automatically
tries to build it from scratch using packages from CRAN. Travis CI is now 
automatically building nprcmanager without errors or warnings. 
This puts us considerably closer to being able to offer the package via 
CRAN if we want.

#### 20170915
Added a genetic uniqueness plot to the Summary Statistics tab.


#### 20170917
I discovered why the plot of genetic uniqueness scores was not as expected. 
It was from an analysis where 0 was the threshold setting, which I think we 
should consider removing as an option. I have set the default to 3, 
which is what Amando Vinson's paper indicated that ONPRC typically used.

I got the logging system integrated into the package. 
I am using the package futile.logger (funny name). Note the checkbox at the 
bottom of the side panel on the two attached images. 
When the Debug on checkbox is checked (it is not checked by default), 
the application writes to a file in the users home directory named
_nprcmanager.log_. Currently I am only logging events occurring the the 
_server.R_ file, because that is where I tend to have most of my problems 
show up. I have attached an example file in which turned on logging at the 
debug level and then read in a couple of pedigree files.

I added code coverage reports to the automated build system running on 
Travis-CI.org. Currently we only have 5.68% of the code being tested with the 
unit tests I created thus far. I do not know the code well enough to know what 
percentage we will have when I feel comfortable, but I know most of the 
functions have no tests at all.

#### 20170919
We have successfully installed and used _nprcmanager_ on Microsoft (MS) 
Windows 7, MS Windows 10, and MacOS 10.12.6 running R 3.4.1.

#### 20170921
Unit test coverage is about 15%.

## Meeting notes

I met with Jack Kent, Deborah Newman, and Charles Peterson about how to use
genetic data in a gene dropping simulation.

From: R. Mark Sharp <msharp@TxBiomed.org>

Subject: Re: Genetic management of colonies

Date: March 23, 2017 at 5:01:07 PM CDT

To: Jack Kent <jkent@txbiomed.org>

Cc: Deborah Newman <dnewman@txbiomedgenetics.org>, 
Charles Peterson <charlesp@txbiomed.org>

Jack, Debbie, and Charles,

Thank you for meeting with me. I appreciate your help in clarifying the 
various stages of the issues associated with providing genetic management 
guidance via simulation with partially known MHC data. 

1. I will look more closely into what will be needed to add genetic 
information and gene frequency information into the gene dropping routines 
in the kinship2 package. The initial implementation will use only a single 
locus, which should be a sufficient for MHC data

	A. First step will be to ensure we get expected proportions of genes in 
	children of individuals when we provide genetic data to all founders using 
	a gene frequency based algorithm.

	B. Second step will be to ensure we get expected proportions of genes in 
	children when all parental generations have fully known genotypes.

	C. Third step will be to ensure we get expected proportions of genes in 
	children when parental generations have partially known genotypes and all 
	other genotypes are uninformative.

	D. Fourth step will be to ensure we get expected proportions of genes in 
	children when parental generations have partially known genotypes and 
	remaining parental genes are determined by gene frequency.

2. We will not worry about getting rid of genes in the pedigrees since colony 
managers need only not breed animals that carry unwanted alleles.

3. I will produce a small number of pedigree drawings using the data Debbie 
has provided to get feedback on preferences and additional requirements.

4. I will not deal with paternity and maternity issues within the pedigrees 
at this time.

Thanks again.

