<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ nprcmanager</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Function reference" />



<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nprcmanager</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.3.23</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Development_Plans.html">Development Plans</a>
    </li>
    <li>
      <a href="../articles/manual.html">Genetic Management Tools Manual</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/rmsharp/nprcmanager">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-data-objects" class="hasAnchor"><a href="#section-data-objects" class="anchor"></a>Data objects</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="baboonBreeders.html">baboonBreeders</a></code> </p>
        </td>
        <td><p>baboonBreeders is a list of 29 baboon IDs that are potential breeders</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="baboonGvReport.html">baboonGvReport</a></code> </p>
        </td>
        <td><p>baboonGvReport is a genetic value report</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="baboonPed.html">baboonPed</a></code> </p>
        </td>
        <td><p>baboonPed is a dataframe with 277 rows and 6 columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="finalRpt.html">finalRpt</a></code> </p>
        </td>
        <td><p>finalRpt is a list object created from the list object <em>rpt</em> prepared
by <code>reportGV</code>. It is created inside <code>orderReport</code>. This version
is at the state just prior to calling <code>rankSubjects</code> inside
<code>orderReport</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lacy1989Ped.html">lacy1989Ped</a></code> </p>
        </td>
        <td><p>lacy1989Ped small hypothetical pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="lacy1989PedAlleles.html">lacy1989PedAlleles</a></code> </p>
        </td>
        <td><p>lacy1989PedAlleles is a dataframe produced by <code>geneDrop</code> on
<code>lacy1989Ped</code> with 5000 iterations.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ped1Alleles.html">ped1Alleles</a></code> </p>
        </td>
        <td><p>ped1Alleles is a dataframe created by the geneDrop function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedWithGenotype.html">pedWithGenotype</a></code> </p>
        </td>
        <td><p>pedWithGenotype is a dataframe produced from baboonPed by adding made up
genotypes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="pedWithGenotypeReport.html">pedWithGenotypeReport</a></code> </p>
        </td>
        <td><p>pedWithGenotypeReport is a list containing the output of <code>reportGV</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rpt.html">rpt</a></code> </p>
        </td>
        <td><p>rpt is a dataframe created from the genetic analysis routine
<code>reportGV</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smallPed.html">smallPed</a></code> </p>
        </td>
        <td><p>smallPed is a hypothetical pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="smallPedTree.html">smallPedTree</a></code> </p>
        </td>
        <td><p>smallPedTree is a pedigree tree made from <code>smallPed</code></p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-codebook" class="hasAnchor"><a href="#section-codebook" class="anchor"></a>Codebook</h2>
          <p class="section-desc"><p>Brief codebook</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="nprcmanager.html">nprcmanager</a></code> </p>
        </td>
        <td><p>Genetic Value Analysis Functions</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-primary-interactive-functions" class="hasAnchor"><a href="#section-primary-interactive-functions" class="anchor"></a>Primary interactive functions</h2>
          <p class="section-desc"><p>Funtions designed for interactive use</p></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="addBackSecondParents.html">addBackSecondParents()</a></code> </p>
        </td>
        <td><p>Add back single parents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addGenotype.html">addGenotype()</a></code> </p>
        </td>
        <td><p>Add genotype data to pedigree file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addParents.html">addParents()</a></code> </p>
        </td>
        <td><p>Add parents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addUIds.html">addUIds()</a></code> </p>
        </td>
        <td><p>Eliminates partial parentage situations by adding unique placeholder
IDs for the unknown parent.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="agePyramidPlot.html">agePyramidPlot()</a></code> </p>
        </td>
        <td><p>Form age pyramid plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assignAlleles.html">assignAlleles()</a></code> </p>
        </td>
        <td><p>Assign parent alleles randomly</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcA.html">calcA()</a></code> </p>
        </td>
        <td><p>Calculates <code>a</code>, the number of an individual's alleles that are rare in
each simulation.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcAge.html">calcAge()</a></code> </p>
        </td>
        <td><p>Calculate animal ages.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFE.html">calcFE()</a></code> </p>
        </td>
        <td><p>Calculates founder Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFEFG.html">calcFEFG()</a></code> </p>
        </td>
        <td><p>Calculates Founder Equivalents and Founder Genome Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFG.html">calcFG()</a></code> </p>
        </td>
        <td><p>Calculates Founder Genome Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcGU.html">calcGU()</a></code> </p>
        </td>
        <td><p>Calculates genome uniqueness for each ID that is part of the population.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcRetention.html">calcRetention()</a></code> </p>
        </td>
        <td><p>Calculates Allelic Retention</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkGenotypeFile.html">checkGenotypeFile()</a></code> </p>
        </td>
        <td><p>Check genotype file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkParentAge.html">checkParentAge()</a></code> </p>
        </td>
        <td><p>Check parent age</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="correctParentSex.html">correctParentSex()</a></code> </p>
        </td>
        <td><p>Sets sex for animals listed as either a sire or dam.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="countLoops.html">countLoops()</a></code> </p>
        </td>
        <td><p>Count the number of loops in a pedigree tree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createPedTree.html">createPedTree()</a></code> </p>
        </td>
        <td><p>Create a pedigree tree (PedTree).</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="finalRpt.html">finalRpt</a></code> </p>
        </td>
        <td><p>finalRpt is a list object created from the list object <em>rpt</em> prepared
by <code>reportGV</code>. It is created inside <code>orderReport</code>. This version
is at the state just prior to calling <code>rankSubjects</code> inside
<code>orderReport</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findGeneration.html">findGeneration()</a></code> </p>
        </td>
        <td><p>Determines the generation number for each id.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findLoops.html">findLoops()</a></code> </p>
        </td>
        <td><p>Find loops in a pedigree tree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findPedigreeNumber.html">findPedigreeNumber()</a></code> </p>
        </td>
        <td><p>Determines the generation number for each id.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fixGenotypeCols.html">fixGenotypeCols()</a></code> </p>
        </td>
        <td><p>Reformat names of observed genotype columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneDrop.html">geneDrop()</a></code> </p>
        </td>
        <td><p>Performs a gene drop simulation based on the provided pedigree information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAncestors.html">getAncestors()</a></code> </p>
        </td>
        <td><p>Recursively create a character vector of ancestors for an individual ID.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getBreederPed.html">getBreederPed()</a></code> </p>
        </td>
        <td><p>Get pedigree based on list of breeders</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDatedFilename.html">getDatedFilename()</a></code> </p>
        </td>
        <td><p>Returns a character vector with an file name having the date prepended.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDemographics.html">getDemographics()</a></code> </p>
        </td>
        <td><p>Get demographic data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGVGenotype.html">getGVGenotype()</a></code> </p>
        </td>
        <td><p>Get Genetic Valuie Genotype data structure for reportGV function.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGVPopulation.html">getGVPopulation()</a></code> </p>
        </td>
        <td><p>Get the population of interest for the Genetic Value analysis.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLkDirectAncestors.html">getLkDirectAncestors()</a></code> </p>
        </td>
        <td><p>Get the direct ancestors of selected animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLkDirectRelatives.html">getLkDirectRelatives()</a></code> </p>
        </td>
        <td><p>Get the direct ancestors of selected animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPedMaxAge.html">getPedMaxAge()</a></code> </p>
        </td>
        <td><p>Get the maximum age of live animals in the pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPossibleCols.html">getPossibleCols()</a></code> </p>
        </td>
        <td><p>Get possible column names for a studbook.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPyramidPlot.html">getPyramidPlot()</a></code> </p>
        </td>
        <td><p>Creates a pyramid plot of the pedigree provided.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getRequiredCols.html">getRequiredCols()</a></code> </p>
        </td>
        <td><p>Get required column names for a studbook.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSiteInfo.html">getSiteInfo()</a></code> </p>
        </td>
        <td><p>Get site information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kinship.html">kinship()</a></code> </p>
        </td>
        <td><p>Generates a kinship matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qcStudbook.html">qcStudbook()</a></code> </p>
        </td>
        <td><p>Quality Control for the Studbook or pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rankSubjects.html">rankSubjects()</a></code> </p>
        </td>
        <td><p>Ranks animals based on genetic value.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reformatKinship.html">reformatKinship()</a></code> </p>
        </td>
        <td><p>Reformats a kinship matrix into a long-format table.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeDuplicates.html">removeDuplicates()</a></code> </p>
        </td>
        <td><p>Remove duplicate records from pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeUninformativeFounders.html">removeUninformativeFounders()</a></code> </p>
        </td>
        <td><p>Remove uninformative founders.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runManager.html">runManager()</a></code> </p>
        </td>
        <td><p>Allows running <code>shiny</code> application with <code>nprcmanager::runManager()</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trimPedigree.html">trimPedigree()</a></code> </p>
        </td>
        <td><p>Trim pedigree to ancestors of provided group removing uninformative
ancestors.</p></td>
      </tr>
    </tbody><tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      <tr>
        
        <td>
          <p><code><a href="addBackSecondParents.html">addBackSecondParents()</a></code> </p>
        </td>
        <td><p>Add back single parents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addGenotype.html">addGenotype()</a></code> </p>
        </td>
        <td><p>Add genotype data to pedigree file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addParents.html">addParents()</a></code> </p>
        </td>
        <td><p>Add parents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="addUIds.html">addUIds()</a></code> </p>
        </td>
        <td><p>Eliminates partial parentage situations by adding unique placeholder
IDs for the unknown parent.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="agePyramidPlot.html">agePyramidPlot()</a></code> </p>
        </td>
        <td><p>Form age pyramid plot</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="allTrueNoNA.html">allTrueNoNA()</a></code> </p>
        </td>
        <td><p>Returns <code>TRUE</code> if every member of the vector is <code>TRUE</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="alleleFreq.html">alleleFreq()</a></code> </p>
        </td>
        <td><p>Calculates the count of each allele in the provided vector.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="assignAlleles.html">assignAlleles()</a></code> </p>
        </td>
        <td><p>Assign parent alleles randomly</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcA.html">calcA()</a></code> </p>
        </td>
        <td><p>Calculates <code>a</code>, the number of an individual's alleles that are rare in
each simulation.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcAge.html">calcAge()</a></code> </p>
        </td>
        <td><p>Calculate animal ages.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFE.html">calcFE()</a></code> </p>
        </td>
        <td><p>Calculates founder Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFEFG.html">calcFEFG()</a></code> </p>
        </td>
        <td><p>Calculates Founder Equivalents and Founder Genome Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcFG.html">calcFG()</a></code> </p>
        </td>
        <td><p>Calculates Founder Genome Equivalents</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcGU.html">calcGU()</a></code> </p>
        </td>
        <td><p>Calculates genome uniqueness for each ID that is part of the population.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="calcRetention.html">calcRetention()</a></code> </p>
        </td>
        <td><p>Calculates Allelic Retention</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkGenotypeFile.html">checkGenotypeFile()</a></code> </p>
        </td>
        <td><p>Check genotype file</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="checkParentAge.html">checkParentAge()</a></code> </p>
        </td>
        <td><p>Check parent age</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="chooseAlleles.html">chooseAlleles()</a></code> </p>
        </td>
        <td><p>Combines two vectors of alleles by randomly selecting one allele
or the other at each position.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="chooseAllelesChar.html">chooseAllelesChar()</a></code> </p>
        </td>
        <td><p>Combines two vectors of alleles when alleles are character vectors.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="chooseDate.html">chooseDate()</a></code> </p>
        </td>
        <td><p>Choose date based on <code>earlier</code> flag.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertAncestry.html">convertAncestry()</a></code> </p>
        </td>
        <td><p>Converts the ancestry information to a standardized code</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertDate.html">convertDate()</a></code> </p>
        </td>
        <td><p>Converts date columns formatted as characters to be of type datetime</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertRelationships.html">convertRelationships()</a></code> </p>
        </td>
        <td><p>Converts pairwise kinship values to a relationship category descriptor.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertSexCodes.html">convertSexCodes()</a></code> </p>
        </td>
        <td><p>Converts sex indicator for an individual to a standardized codes.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="convertStatusCodes.html">convertStatusCodes()</a></code> </p>
        </td>
        <td><p>Converts status indicators to a Standardized code</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="correctParentSex.html">correctParentSex()</a></code> </p>
        </td>
        <td><p>Sets sex for animals listed as either a sire or dam.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="countFirstOrder.html">countFirstOrder()</a></code> </p>
        </td>
        <td><p>Count first-order relatives.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="countLoops.html">countLoops()</a></code> </p>
        </td>
        <td><p>Count the number of loops in a pedigree tree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="createPedTree.html">createPedTree()</a></code> </p>
        </td>
        <td><p>Create a pedigree tree (PedTree).</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fillBins.html">fillBins()</a></code> </p>
        </td>
        <td><p>fillBins Fill bins represented by list of two lists <code>males</code> and <code>females</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterAge.html">filterAge()</a></code> </p>
        </td>
        <td><p>Removes kinship values where an animal is less than the minAge</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterKinMatrix.html">filterKinMatrix()</a></code> </p>
        </td>
        <td><p>Filters a kinship matrix to include only the egos listed in 'ids'</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterPairs.html">filterPairs()</a></code> </p>
        </td>
        <td><p>Filters kinship values from a long-format kinship table based on the sexes
 of the two animals involved.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterReport.html">filterReport()</a></code> </p>
        </td>
        <td><p>Filters a genetic value report down to only the specified animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="filterThreshold.html">filterThreshold()</a></code> </p>
        </td>
        <td><p>Filters kinship to remove rows with kinship values less than the specified
threshold</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findGeneration.html">findGeneration()</a></code> </p>
        </td>
        <td><p>Determines the generation number for each id.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findLoops.html">findLoops()</a></code> </p>
        </td>
        <td><p>Find loops in a pedigree tree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findOffspring.html">findOffspring()</a></code> </p>
        </td>
        <td><p>Finds the number of total offspring for each animal in the provided pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="findPedigreeNumber.html">findPedigreeNumber()</a></code> </p>
        </td>
        <td><p>Determines the generation number for each id.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="fixGenotypeCols.html">fixGenotypeCols()</a></code> </p>
        </td>
        <td><p>Reformat names of observed genotype columns</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="geneDrop.html">geneDrop()</a></code> </p>
        </td>
        <td><p>Performs a gene drop simulation based on the provided pedigree information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getAncestors.html">getAncestors()</a></code> </p>
        </td>
        <td><p>Recursively create a character vector of ancestors for an individual ID.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getBreederPed.html">getBreederPed()</a></code> </p>
        </td>
        <td><p>Get pedigree based on list of breeders</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDatedFilename.html">getDatedFilename()</a></code> </p>
        </td>
        <td><p>Returns a character vector with an file name having the date prepended.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getDemographics.html">getDemographics()</a></code> </p>
        </td>
        <td><p>Get demographic data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGVGenotype.html">getGVGenotype()</a></code> </p>
        </td>
        <td><p>Get Genetic Valuie Genotype data structure for reportGV function.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getGVPopulation.html">getGVPopulation()</a></code> </p>
        </td>
        <td><p>Get the population of interest for the Genetic Value analysis.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getIncludeColumns.html">getIncludeColumns()</a></code> </p>
        </td>
        <td><p>Get the superset of columns that can be in a pedigree file.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLkDirectAncestors.html">getLkDirectAncestors()</a></code> </p>
        </td>
        <td><p>Get the direct ancestors of selected animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLkDirectRelatives.html">getLkDirectRelatives()</a></code> </p>
        </td>
        <td><p>Get the direct ancestors of selected animals</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getLogo.html">getLogo()</a></code> </p>
        </td>
        <td><p>Get Logo file name</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMaxAx.html">getMaxAx()</a></code> </p>
        </td>
        <td><p>Get the maximum of the absolute values of the negative (males) and positive
(female) animal counts.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getMinParentAge.html">getMinParentAge()</a></code> </p>
        </td>
        <td><p>Get minimum parent age.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getOffspring.html">getOffspring()</a></code> </p>
        </td>
        <td><p>Get offspring to corresponding animal IDs provided</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getParamDef.html">getParamDef()</a></code> </p>
        </td>
        <td><p>Get parameter definitions from tokens found in configuration file.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getParents.html">getParents()</a></code> </p>
        </td>
        <td><p>Get parents to corresponding animal IDs provided</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPedMaxAge.html">getPedMaxAge()</a></code> </p>
        </td>
        <td><p>Get the maximum age of live animals in the pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPossibleCols.html">getPossibleCols()</a></code> </p>
        </td>
        <td><p>Get possible column names for a studbook.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getProbandPedigree.html">getProbandPedigree()</a></code> </p>
        </td>
        <td><p>Gets pedigree to ancestors of provided group leaving uninformative ancestors.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPyramidAgeDist.html">getPyramidAgeDist()</a></code> </p>
        </td>
        <td><p>Get the age distribution for the pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getPyramidPlot.html">getPyramidPlot()</a></code> </p>
        </td>
        <td><p>Creates a pyramid plot of the pedigree provided.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getRequiredCols.html">getRequiredCols()</a></code> </p>
        </td>
        <td><p>Get required column names for a studbook.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getSiteInfo.html">getSiteInfo()</a></code> </p>
        </td>
        <td><p>Get site information</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="getTokenList.html">getTokenList()</a></code> </p>
        </td>
        <td><p>Returns first right and left space trimmed token from first character vector
element.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="groupAddAssign.html">groupAddAssign()</a></code> </p>
        </td>
        <td><p>Add animals to an existing breeding or form groups group:</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="hasGenotype.html">hasGenotype()</a></code> </p>
        </td>
        <td><p>Check for genotype data in dataframe</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="headerDisplayNames.html">headerDisplayNames()</a></code> </p>
        </td>
        <td><p>Convert internal column names to display or header names.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="isEmpty.html">isEmpty()</a></code> </p>
        </td>
        <td><p>Is vector empty or all NA values.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="kinship.html">kinship()</a></code> </p>
        </td>
        <td><p>Generates a kinship matrix.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeCEPH.html">makeCEPH()</a></code> </p>
        </td>
        <td><p>Make a CEPH-style pedigree for each id</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeRelationClasseTable.html">makeRelationClasseTable()</a></code> </p>
        </td>
        <td><p>Make relation classes table from <code>kin</code> dataframe.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makeRoundUp.html">makeRoundUp()</a></code> </p>
        </td>
        <td><p>Round up the provided integer vector <code>int</code> according to the
<code>modulas</code>.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="makesLoop.html">makesLoop()</a></code> </p>
        </td>
        <td><p><code>makesLoop</code> tests for a common ancestor.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="meanKinship.html">meanKinship()</a></code> </p>
        </td>
        <td><p>Calculates the mean kinship for each animal in a kinship matrix</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="offspringCounts.html">offspringCounts()</a></code> </p>
        </td>
        <td><p>Finds the total number of offspring for each animal in the pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="orderReport.html">orderReport()</a></code> </p>
        </td>
        <td><p>Order the results of the genetic value analysis for use in a report.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="qcStudbook.html">qcStudbook()</a></code> </p>
        </td>
        <td><p>Quality Control for the Studbook or pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rankSubjects.html">rankSubjects()</a></code> </p>
        </td>
        <td><p>Ranks animals based on genetic value.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rebindFill.html">rebindFill()</a></code> </p>
        </td>
        <td><p>Append the rows of one dataframe to another.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reformatKinship.html">reformatKinship()</a></code> </p>
        </td>
        <td><p>Reformats a kinship matrix into a long-format table.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeDuplicates.html">removeDuplicates()</a></code> </p>
        </td>
        <td><p>Remove duplicate records from pedigree</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="removeUninformativeFounders.html">removeUninformativeFounders()</a></code> </p>
        </td>
        <td><p>Remove uninformative founders.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="reportGV.html">reportGV()</a></code> </p>
        </td>
        <td><p>Generates a genetic value report for a provided pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resetGroup.html">resetGroup()</a></code> </p>
        </td>
        <td><p>Update or add the "group" field of a Pedigree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="resetPopulation.html">resetPopulation()</a></code> </p>
        </td>
        <td><p>Population designation function</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="runManager.html">runManager()</a></code> </p>
        </td>
        <td><p>Allows running <code>shiny</code> application with <code>nprcmanager::runManager()</code></p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="setExit.html">setExit()</a></code> </p>
        </td>
        <td><p>Sets the exit date, if there is no exit column in the table</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="toCharacter.html">toCharacter()</a></code> </p>
        </td>
        <td><p>Force dataframe columns to character</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="trimPedigree.html">trimPedigree()</a></code> </p>
        </td>
        <td><p>Trim pedigree to ancestors of provided group removing uninformative
ancestors.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="unknown2NA.html">unknown2NA()</a></code> </p>
        </td>
        <td><p>Removing IDs having "UNKNOWN" regardless of case</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#section-data-objects">Data objects</a></li>
      <li><a href="#section-codebook">Codebook</a></li>
      <li><a href="#section-primary-interactive-functions">Primary interactive functions</a></li>
      <li><a href="#section-all-functions">All functions</a></li>
    </ul>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Michael Raboin, R. Mark Sharp.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

