<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />



<meta name="progressive" content="false" />
<meta name="allow-skip" content="false" />

<title>Interactive Use of nprcmanager</title>


<!-- highlightjs -->
<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>

<!-- taken from https://github.com/rstudio/rmarkdown/blob/67b7f5fc779e4cfdfd0f021d3d7745b6b6e17149/inst/rmd/h/default.html#L296-L362 -->
<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("section-TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>
<!-- end tabsets -->



</head>

<body>



<div class="pageContent band">
<div class="bandContent page">

<div class="topics">

<div id="section-introduction" class="section level2">
<h2>Introduction</h2>
<p>This tutorial demonstrates the major functions used within the Shiny application provided by the <strong>nprcmanager</strong> package and provides sufficient insight into those functions that they may be used independently.</p>
<p>This tutorial is primarily directed toward someone with experience using R who wants to better understand how the Shiny application works or to perform some actions not directly supported by the Shiny application.</p>
<p>Please provide any comments, questions, or bug reports through the GitHub issue tracker at .</p>
</div>
<div id="section-installation-and-help" class="section level2">
<h2>Installation and Help</h2>
<p>You can install <strong>nprcmanager</strong> from GitHub with the following code.</p>
<pre class="r"><code>install.packages(&quot;devtools&quot;)
devtools::install_github(&quot;rmsharp/nprcmanager&quot;)</code></pre>
<p>All missing dependencies should be automatically installed.</p>
<p>You can get help from the <strong>R</strong> console with</p>
<pre class="r"><code>?nprcmanager</code></pre>
<p><strong>The help provided by this (<em>nprcmanager.R</em>) needs to be more complete and include links to the tutorials.</strong></p>
</div>
<div id="section-reading-in-a-pedigree" class="section level2">
<h2>Reading in a Pedigree</h2>
<p>A pedigrees can be imported using either Excel worksheets or text files that contain all of the pedigree information or using either Excel worksheets or text files that contain a list of focal animals with the remainder of the pedigree information is pulled in through the LabKey API.</p>
<p>This tutorial will use a pedigree file that can be created using the <strong>makeExamplePedigreeFile</strong> function as shown below. The function <strong>makeExamplePedigreeFile</strong> both saves a file and returns the full path name to the saved file, which we are saving into the variable <em>pedigreeFile</em>. Note: the user will select where to store the file.</p>
<pre class="r"><code>library(nprcmanager)
pedigreeFile &lt;- makeExamplePedigreeFile()</code></pre>
<p>This writes <em>ExamplePedigree.csv</em> to a place you select within your file system.</p>
<p>You use the file name provided by the <strong>makeExamplePedigreeFile</strong> function to tell <strong>read.table</strong> what file to read.</p>
<pre class="r"><code>breederPedCsv &lt;- read.table(pedigreeFile, sep = &quot;,&quot;, header = TRUE,
                            stringsAsFactors = FALSE)</code></pre>
<p>Note the number of rows read. Each row represents an individual within the pedigree.</p>
<pre class="r"><code>nrow(breederPedCsv)</code></pre>
<pre><code>## [1] 3694</code></pre>
<p>The next step is to put the information read from the file into a pedigree object. This is done with the <strong>qcStudbook</strong> function, which examines the file contents and tests for common pedigree errors.</p>
<p>You can see the errors that can be detected by <strong>qcStudbook</strong> by returning the empty error list with <strong>getEmptyErrorLst()</strong>. We are not showing the output of the function call now because later in this tutorial we will explore errors in more depth.</p>
<p><strong>qcStudbook</strong> can take four arguments <em>sb</em>, <em>minParentAge</em> (in years), <em>reportChanges</em>, and <em>reportErrors</em>. However, all but <em>sb</em> have default values and only the <em>sb</em> argument is required.</p>
<p>It is prudent to ensure that parents are at least of breeding age, which is species specific. I have used a <em>minParentAge</em> of 2 years.<a href="#section-fn1" class="footnote-ref" id="section-fnref1"><sup>1</sup></a></p>
<pre class="r"><code>breederPed &lt;- qcStudbook(breederPedCsv, minParentAge = 2)</code></pre>
<p>If <strong>qcStudbook</strong> reports an error, change the call by adding the <strong>reportErrors</strong> argument set to <strong>TRUE</strong> and examine the returned object. More on this is presented in the <strong>Pedigree Errors</strong> section.</p>
</div>
<div id="section-identifying-focal-animals" class="section level2">
<h2>Identifying Focal Animals</h2>
<p>You may want to focus your work on a <em>focal</em> group of animals. This can be done by reading in a list of animal IDs that make up the <em>focal</em> group and use that list to update the pedigree. Alternatively you can created a list of animal IDs based on criteria you have selected.</p>
<p>For example, to select living animals at the facility with at least one parent, the following can be used.</p>
<pre class="r"><code>focalAnimals &lt;- breederPed$id[!(is.na(breederPed$sire) &amp; 
                                  is.na(breederPed$dam)) &amp;
                                is.na(breederPed$exit)]
print(stri_c(&quot;There are &quot;, length(focalAnimals), 
             &quot; animals in the vector _focalAnimals_.&quot;))</code></pre>
<p>[1] “There are 327 animals in the vector <em>focalAnimals</em>.”</p>
<p>As can be seen, these animals have at least one parent and have not left the facility.</p>
<pre class="r"><code>breederPed[breederPed$id %in% 
             focalAnimals, c(&quot;id&quot;, &quot;sire&quot;, &quot;dam&quot;, &quot;exit&quot;)][1:10, ]</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["id"],"name":[1],"type":["chr"],"align":["left"]},{"label":["sire"],"name":[2],"type":["chr"],"align":["left"]},{"label":["dam"],"name":[3],"type":["chr"],"align":["left"]},{"label":["exit"],"name":[4],"type":["date"],"align":["right"]}],"data":[{"1":"01QRQ4","2":"VDBGDP","3":"TH7HTY","4":"<NA>","_rn_":"1669"},{"1":"CLSVU6","2":"ULV9M7","3":"SUFWJI","4":"<NA>","_rn_":"1743"},{"1":"1SPLS8","2":"U9APLW","3":"142GKP","4":"<NA>","_rn_":"1887"},{"1":"5IAFMK","2":"U4YSS5","3":"WVE6Y4","4":"<NA>","_rn_":"1934"},{"1":"HLQ9SY","2":"UI3RFL","3":"VEWC1E","4":"<NA>","_rn_":"2072"},{"1":"XFWVVX","2":"U3MFSZ","3":"L4LM1F","4":"<NA>","_rn_":"2234"},{"1":"6X6BG9","2":"ENI6HX","3":"IUF0HC","4":"<NA>","_rn_":"2337"},{"1":"B228Q6","2":"UEUIRJ","3":"CBSIAA","4":"<NA>","_rn_":"2377"},{"1":"B2CKHA","2":"ENI6HX","3":"WBFBR5","4":"<NA>","_rn_":"2378"},{"1":"BCJJKN","2":"UA379T","3":"JPVAT3","4":"<NA>","_rn_":"2383"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>We indicate that these are the animals of interest by using the <strong>setPopulation</strong> function. This function simply sets a column named <em>population</em><a href="#section-fn2" class="footnote-ref" id="section-fnref2"><sup>2</sup></a> to the logical value of <strong>TRUE</strong> if the row represents an animal in the list and <strong>FALSE</strong> otherwise.</p>
<p>The first line of code below sets the <em>population</em> column and the second counts the number of rows where the value was set to <strong>TRUE</strong>.</p>
<pre class="r"><code>breederPed &lt;- setPopulation(ped = breederPed, ids = focalAnimals)
nrow(breederPed[breederPed$population, ])</code></pre>
<pre><code>## [1] 327</code></pre>
<p>The IDs used to populate the <em>population</em> flag can be used to trim the pedigree so that it contains only those individuals who are in the ID list or are ancestors of those individuals.</p>
<pre class="r"><code>trimmedPed &lt;- trimPedigree(focalAnimals, breederPed)
nrow(breederPed); nrow(trimmedPed)</code></pre>
<pre><code>## [1] 3694</code></pre>
<pre><code>## [1] 704</code></pre>
<p>The <strong>trimPedigree</strong> function has the ability to remove those ancestors that do not contribute genetic information. Uninformative founders are those individuals who are parents of only one individual and who have no parental information. (<em>Currently genotypic information is ignored by <strong>trimPedigree</strong></em>).</p>
<pre class="r"><code>trimmedPedInformative &lt;- trimPedigree(focalAnimals, breederPed,
                                      removeUninformative = TRUE)
nrow(trimmedPedInformative)</code></pre>
<pre><code>## [1] 509</code></pre>
<p>We can find all of the animals that are in the trimmed pedigree but are not focal animals.</p>
<pre class="r"><code>nonfocalInTrimmedPed &lt;- trimmedPed$id[!trimmedPed$id %in% focalAnimals]
length(nonfocalInTrimmedPed)</code></pre>
<pre><code>## [1] 377</code></pre>
<p>We can see which of these 377 are and are not parents. We will first make sure we have all of the parents by getting our list of parents from the entire pedigree. We then demonstrate that they are all in the trimmed pedigree.</p>
<pre class="r"><code>allFocalParents &lt;- c(breederPed$sire[breederPed$id %in% focalAnimals], 
                       breederPed$dam[breederPed$id %in% focalAnimals])
trimmedFocalParents &lt;- c(trimmedPed$sire[trimmedPed$id %in% focalAnimals], 
                       trimmedPed$dam[trimmedPed$id %in% focalAnimals])
all.equal(allFocalParents, trimmedFocalParents) # Are the IDs the same?</code></pre>
<pre><code>## [1] TRUE</code></pre>
<p>However, not all of the animals in the trimmed pedigree are either the focal animals or their parents. They are more distant ancestors as we will show.</p>
<pre class="r"><code>notFocalNotParent &lt;- trimmedPed$id[!trimmedPed$id %in% c(focalAnimals, allFocalParents)]
length(notFocalNotParent)</code></pre>
<pre><code>## [1] 187</code></pre>
<p>Since the trimming process is supposed to retain the focal animals and their ancestors, we will leave it as an exercise for you to demonstrate that at least some of the remaining animals are grandparents of the focal animals. <em>Hint: there are 490 grandparents in both the trimmed and the complete pedigree</em>.</p>
<p>As you can see from the number of rows in the full pedigree (3694) versus the trimmed pedigree (704), trimmed pedigrees can be much smaller. Of the additional 377 animals, 182 provide genetic information while the others (195) are genetically uninformative.</p>
<p>As is shown below only 4 (0ZX29Q, 1QBKW9, 5PWJ0G, and Y3CJ5A) living animals are still in the colony but are not in the trimmed pedigree.<a href="#section-fn3" class="footnote-ref" id="section-fnref3"><sup>3</sup></a></p>
<pre class="r"><code>unknownBirth &lt;- breederPed$id[is.na(breederPed$birth)]
knownExit &lt;- breederPed$id[ !is.na(breederPed$exit)]
unknownBirthKnownExit &lt;- breederPed$id[is.na(breederPed$birth) | !is.na(breederPed$exit)]
knownPed &lt;- breederPed[!breederPed$id %in% unknownBirthKnownExit, ]
otherIds &lt;- knownPed$id[!knownPed$id %in% trimmedPed$id[is.na(trimmedPed$exit)]]
print(stri_c(&quot;The living animal in the pedigree that are not in the trimmed &quot;,
             &quot;pedigree are &quot;, get_and_or_list(otherIds), &quot;.&quot;))</code></pre>
<p>[1] “The living animal in the pedigree that are not in the trimmed pedigree are 0ZX29Q, 1QBKW9, 5PWJ0G, and Y3CJ5A.”</p>
</div>
<div id="section-age-sex-pyramid-plot" class="section level2">
<h2>Age Sex Pyramid Plot</h2>
<p>You can examine the population structure using an age-sex pyramid plot with a single function. We will limit our view to just the focal animals and their living relatives. This is appropriate for colony management because in addition to the genetic diversity we seek, we have to remain cognacent of the age and sex distributions within the colonies we manage.</p>
<pre class="r"><code>getPyramidPlot(ped = trimmedPed[is.na(trimmedPed$exit), ])</code></pre>
<p><img src="interactive_learnr_files/figure-html/plot-focal-age-sex-pyramid-1.png" width="624" /></p>
</div>
<div id="section-genetic-value-analysis" class="section level2">
<h2>Genetic Value Analysis</h2>
<p>Your genetic value analysis must be carefully performed. The next three commands set up the entire pedigree for analysis. The first of these three commands set all of the pedigree members to be part of the population of interest by setting the <em>population</em> column to <strong>TRUE</strong> for all individuals.</p>
<pre class="r"><code>ped &lt;- setPopulation(breederPed, NULL)</code></pre>
<p>Note that a new pedigree object (<strong>ped</strong>) is being created.</p>
<pre class="r"><code>probands &lt;- ped$id[ped$population]
ped &lt;- trimPedigree(probands, ped, removeUninformative = FALSE,
                    addBackParents = FALSE)</code></pre>
<p>The arguments to <strong>reportGV</strong> are all optional except for <em>ped</em>, but you may often want to non-default values.</p>
<ul>
<li><p><strong>ped</strong> Pedigree information in data.frame format</p></li>
<li><p><strong>guIter</strong> Integer indicating the number of iterations for the gene-drop analysis. Default is 5000 iterations</p></li>
<li><p><strong>guThresh</strong> Integer indicating the threshold number of animals for defining a unique allele. Default considers an allele “unique” if it is found in only 1 animal.</p></li>
<li><p><strong>pop</strong> Character vector with animal IDs to consider as the population of interest. The default is NULL.</p></li>
<li><p><strong>byID</strong> Logical variable of length 1 that is passed through to eventually be used by alleleFreq(), which calculates the count of each allele in the provided vector. If byID is TRUE and ids are provided, the function will only count the unique alleles for an individual (homozygous alleles will be counted as 1).</p></li>
</ul>
<pre class="r"><code>geneticValue &lt;- reportGV(ped, guIter = 500,
                         guThresh = 3,
                         byID = TRUE,
                         updateProgress = NULL)
summary(geneticValue)</code></pre>
<pre><code>## The genetic value report 
## Individuals in Pedigree: 3694 
## Male Founders: 141
## Female Founders: 122
## Total Founders: 263 
## Founder Equivalents: 241.84 
## Founder Genome Equivalents: 163.84 
## Live Offspring: 4052 
## High Value Individuals: 2961 
## Low Value Individuals: 733</code></pre>
<p>What happens if we limit our analysis to the trimmed pedigree? Remember that the trimmed pedigree still contains all of the genetic information that the full pedigree has for the focal animals.</p>
<pre class="r"><code>trimmedGeneticValue &lt;- reportGV(trimmedPed, guIter = 500,
                         guThresh = 3,
                         byID = TRUE,
                         updateProgress = NULL)
summary(trimmedGeneticValue)</code></pre>
<pre><code>## The genetic value report 
## Individuals in Pedigree: 327 
## Male Founders: 3
## Female Founders: 17
## Total Founders: 20 
## Founder Equivalents: 109.67 
## Founder Genome Equivalents: 47.87 
## Live Offspring: 321 
## High Value Individuals: 228 
## Low Value Individuals: 99</code></pre>
<p>It is clear that limiting your analysis to the animals of interest reduces the effort required to examine the animals of interest.</p>
<div id="section-detailed-look-at-the-genetic-value-report-object" class="section level3">
<h3>Detailed look at the Genetic Value Report object</h3>
<p>The names of the object within the genetic value report object (<em>trimmedGeneticValue</em>) can be listed as shown in the next line of code.</p>
<pre class="r"><code>names(trimmedGeneticValue)</code></pre>
<pre><code>## [1] &quot;report&quot;         &quot;kinship&quot;        &quot;gu&quot;             &quot;fe&quot;            
## [5] &quot;fg&quot;             &quot;maleFounders&quot;   &quot;femaleFounders&quot; &quot;total&quot;</code></pre>
<p>The <em>report</em> object (an R dataframe) can in-turn be examined.</p>
<pre class="r"><code>names(trimmedGeneticValue$report) ## column names</code></pre>
<pre><code>##  [1] &quot;id&quot;              &quot;sex&quot;             &quot;age&quot;            
##  [4] &quot;birth&quot;           &quot;exit&quot;            &quot;population&quot;     
##  [7] &quot;origin&quot;          &quot;indivMeanKin&quot;    &quot;zScores&quot;        
## [10] &quot;gu&quot;              &quot;totalOffspring&quot;  &quot;livingOffspring&quot;
## [13] &quot;value&quot;           &quot;rank&quot;</code></pre>
<pre class="r"><code>nrow(trimmedGeneticValue$report) ## Number of rows</code></pre>
<pre><code>## [1] 327</code></pre>
<p>The report is more conveniently used as a separate object. The next section of code rounds some of the numerica values and converts all columns to characters for display as a table where only the first 10 lines are included.</p>
<pre class="r"><code>rpt &lt;- trimmedGeneticValue[[&quot;report&quot;]]
rpt$indivMeanKin &lt;- round(rpt$indivMeanKin, 5)
rpt$zScores &lt;- round(rpt$zScores, 2)
rpt$gu &lt;- round(rpt$gu, 5)
rpt &lt;- toCharacter(rpt)
names(rpt) &lt;- headerDisplayNames(names(rpt))
knitr::kable(rpt[1:10, ]) # needs more work for display purposes.</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">Ego ID</th>
<th align="left">Sex</th>
<th align="right">Age (in years)</th>
<th align="left">Birth Date</th>
<th align="left">Exit Date</th>
<th align="left">Breeding Colony Member</th>
<th align="left">Origin</th>
<th align="right">Individual Mean Kinship</th>
<th align="right">Z-score (Mean Kinship)</th>
<th align="right">Genome Uniqueness (%)</th>
<th align="right">Total Offspring</th>
<th align="right">Living Offspring</th>
<th align="left">Value Designation</th>
<th align="right">Rank</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">KZM9RB</td>
<td align="left">M</td>
<td align="right">30.1</td>
<td align="left">1989-05-03</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00329</td>
<td align="right">-1.90</td>
<td align="right">89.1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">CLSVU6</td>
<td align="left">F</td>
<td align="right">23.9</td>
<td align="left">1995-08-02</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00287</td>
<td align="right">-1.97</td>
<td align="right">86.7</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="left">High Value</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">1SPLS8</td>
<td align="left">F</td>
<td align="right">7.9</td>
<td align="left">2011-07-26</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00373</td>
<td align="right">-1.83</td>
<td align="right">81.9</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">WK89I9</td>
<td align="left">F</td>
<td align="right">21.1</td>
<td align="left">1998-05-26</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00582</td>
<td align="right">-1.49</td>
<td align="right">76.2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">3MMZD4</td>
<td align="left">M</td>
<td align="right">12.2</td>
<td align="left">2007-03-24</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00536</td>
<td align="right">-1.57</td>
<td align="right">74.2</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">5</td>
</tr>
<tr class="even">
<td align="left">01QRQ4</td>
<td align="left">F</td>
<td align="right">18.2</td>
<td align="left">2001-04-04</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00373</td>
<td align="right">-1.83</td>
<td align="right">72.8</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">IZDV8K</td>
<td align="left">M</td>
<td align="right">7.7</td>
<td align="left">2011-09-29</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00480</td>
<td align="right">-1.66</td>
<td align="right">72.6</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">8YP6PA</td>
<td align="left">M</td>
<td align="right">5.0</td>
<td align="left">2014-07-04</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00485</td>
<td align="right">-1.65</td>
<td align="right">71.7</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">CHK1ZX</td>
<td align="left">F</td>
<td align="right">7.8</td>
<td align="left">2011-09-06</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00481</td>
<td align="right">-1.66</td>
<td align="right">69.0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="left">High Value</td>
<td align="right">9</td>
</tr>
<tr class="even">
<td align="left">A6A1M1</td>
<td align="left">M</td>
<td align="right">20.0</td>
<td align="left">1999-06-08</td>
<td align="left">NA</td>
<td align="left">TRUE</td>
<td align="left"></td>
<td align="right">0.00577</td>
<td align="right">-1.50</td>
<td align="right">68.9</td>
<td align="right">2</td>
<td align="right">1</td>
<td align="left">High Value</td>
<td align="right">10</td>
</tr>
</tbody>
</table>
<p>We start the next lines of code by getting a fresh copy of the genetic value report since we changed all of the numeric values to characters in the last section to print the table. These lines demonstrate one way of extracting the component objects from the <em>genetic value</em> object.</p>
<pre class="r"><code>rpt &lt;- trimmedGeneticValue[[&quot;report&quot;]]
kmat &lt;- trimmedGeneticValue[[&quot;kinship&quot;]]
f &lt;- trimmedGeneticValue[[&quot;total&quot;]]
mf &lt;- trimmedGeneticValue[[&quot;maleFounders&quot;]]
ff &lt;- trimmedGeneticValue[[&quot;femaleFounders&quot;]]
fe &lt;- trimmedGeneticValue[[&quot;fe&quot;]]
fg &lt;- trimmedGeneticValue[[&quot;fg&quot;]]</code></pre>
<p>It is informative to examine the distribution of <em>genetic uniqueness</em>, <em>mean kinship</em>, and <em>z-scores</em> (normalized <em>mean kinship</em> values).</p>
<p>Creation of the boxplot for the <em>genetic uniqueness</em> values is shown below.</p>
<pre class="r"><code>gu &lt;- rpt[, &quot;gu&quot;]
guBox &lt;- ggplot(data.frame(gu = gu), aes(x = &quot;&quot;, y = gu)) +
  geom_boxplot(
    color = &quot;darkblue&quot;,
    fill = &quot;lightblue&quot;,
    notch = TRUE,
    outlier.color = &quot;red&quot;,
    outlier.shape = 1
  ) +
  theme_classic() + geom_jitter(width = 0.2) + coord_flip() +
  ylab(&quot;Score&quot;)  + ggtitle(&quot;Genetic Uniqueness&quot;)
print(guBox)</code></pre>
<p><img src="interactive_learnr_files/figure-html/genetic-uniqueness-boxplot-1.png" width="624" /></p>
<p>Extraction of the individual <em>mean kinship</em> values and their corresponding z-scores is shown in the next code chunk.</p>
<pre class="r"><code>mk &lt;- rpt[, &quot;indivMeanKin&quot;]
zs &lt;- rpt[, &quot;zScores&quot;]</code></pre>
<p>Creation of boxplots for the <em>mean kinship</em> and <em>z-scores</em> is left as an exercise.</p>
</div>
</div>
<div id="section-breeding-group-formation" class="section level2">
<h2>Breeding Group Formation</h2>
<p>The primary purpose of <strong>nprcmanager</strong> is to form breeding groups according to our best information regarding maintaining the genetic characteristics we desire and the realities associated with other animal husbandry needs.</p>
<p>There are several options you must consider when forming groups using <strong>nprcmanager</strong>, which we will examine using code below.</p>
<ul>
<li>Animals used to form groups
<ul>
<li><em>high-value</em>: Randomly select from only high-value animals in genetic value analysis</li>
<li><em>all</em>: Randomly select from all animals in genetic value analysis</li>
<li><em>candidates</em>: Use candidate animals entered below to form groups</li>
</ul></li>
<li>Sex ratio
<ul>
<li>Randomly assign animals without regard to sex</li>
<li>Use a harem structure with one breeding male per group</li>
<li>Specify the sex ratio between 0.5 and 10 (F/M)</li>
</ul></li>
<li>Whether or not to prepopulate (seed) groups with animals of your choice</li>
<li>Number of groups to be formed</li>
<li>Whether or not to ignore females at or above the minimum parent age</li>
<li>Number of simulations used to search for the optimal group makeup</li>
<li>Whether or not kinship coefficents are to be included in results</li>
</ul>
<p>You decisions regarding each of the above options are expressed in a call to the function <strong>groupAddAssign</strong>. A complete description of the function and its arguments is available using the code shown below.</p>
<pre class="r"><code>?groupAddAssign</code></pre>
<p>Below is the descriptions of the function parameters extracted from the documentation near the time this tutorial was prepared.</p>
<ul>
<li><p><strong>candidates</strong> Character vector of IDs of the animals available for use in forming the groups. The animals that may be present in currentGroups are not included within candidates.</p></li>
<li><p><strong>currentGroups</strong> List of character vectors of IDs of animals currently assigned to groups. Defaults to a list with character(0) in each sublist element (one for each group being formed) assuming no groups are prepopulated.</p></li>
<li><p><strong>kmat</strong> Numeric matrix of pairwise kinship values. Rows and columns are named with animal IDs.</p></li>
<li><p><strong>ped</strong> Dataframe that is the ‘Pedigree’. It contains pedigree information including the IDs listed in candidates.</p></li>
<li><p><strong>threshold</strong> Numeric value indicating the minimum kinship level to be considered in group formation. Pairwise kinship below this level will be ignored. The default values is 0.015625.</p></li>
<li><p><strong>ignore</strong> List of character vectors representing the sex combinations to be ignored. If provided, the vectors in the list specify if pairwise kinship should be ignored between certain sexes. Default is to ignore all pairwise kinship between females.</p></li>
<li><p><strong>minAge</strong> Integer value indicating the minimum age to consider in group formation. Pairwise kinships involving an animal of this age or younger will be ignored. Default is 1 year.</p></li>
<li><p><strong>iter</strong> Integer indicating the number of times to perform the random group formation process. Default value is 1000 iterations.</p></li>
<li><p><strong>numGp</strong> Integer value indicating the number of groups that should be formed from the list of IDs. Default is 1.</p></li>
<li><p><strong>updateProgress</strong> Function or NULL. If this function is defined, it will be called during each iteration to update a shiny::Progress object.</p></li>
<li><p><strong>harem</strong> Logical variable when set to TRUE, the formed groups have a single male at least minAge old.</p></li>
<li><p><strong>sexRatio</strong> Numeric value indicating the ratio of females to males x from 0.5 to 20 by increments of 0.5.</p></li>
<li><p><strong>withKin</strong> Logical variable when set to TRUE, the kinship matrix for the group is returned along with the group and score. Defaults to not return the kinship matrix. This maintains compatability with earlier versions.</p></li>
</ul>
<p>We will use the <em>trimmedPed</em> pedigree in our code.</p>
<p>For illustration purposes we are going to keep the numbers of candidates, groups, and iterations fairly small.</p>
<p>We will get first some animal IDs to use for our candidates by selecting animals at least 2 years old at the time this pedigree was sampled (01-01-2015).</p>
<pre class="r"><code>candidates &lt;- trimmedPed$id[
  trimmedPed$birth &lt; as.Date(&quot;2013-01-01&quot;) &amp; !is.na(trimmedPed$birth) &amp;
    is.na(trimmedPed$exit)]
table(trimmedPed$sex[trimmedPed$id %in% candidates])</code></pre>
<pre><code>## 
##   F   M   H   U 
## 184  96   0   0</code></pre>
<p>Our candidates are made up of 184 females and 96 males. The parameters <strong>currentGroups</strong>, <strong>threshold</strong>, <strong>ignore</strong>, <strong>minAge</strong>, <strong>sexRatio</strong>, <strong>withKin</strong>, and <strong>updateProgress</strong> are allowed to take their default values. The setting of the <strong>sexRatio</strong> parameter to 0 is ignored in the following call of the <strong>groupAddAssign</strong> function. This is consistent with the a value of 0 making little since in a breeding colony.</p>
<p>The empty seventh group at the bottom is evidence that all of the canditate animals could be placed in a group without exceeding the default value of 0.015625.</p>
<div id="section-harems" class="section level3">
<h3>Harems</h3>
<p>The following group assignments will be forming harem groups. This is done by setting <strong>harem</strong> to </p>
<pre class="r"><code>haremGrp &lt;- groupAddAssign(candidates = candidates,
                     kmat = trimmedGeneticValue[[&quot;kinship&quot;]],
                     ped = trimmedPed,
                     iter = 100,
                     numGp = 6,
                     harem = TRUE)
haremGrp$group</code></pre>
<pre><code>## [[1]]
##  [1] &quot;CHSCFG&quot; &quot;B1WVCN&quot; &quot;72LYDE&quot; &quot;13B1QL&quot; &quot;WI38KZ&quot; &quot;AIHJ8Z&quot; &quot;EMV4P6&quot;
##  [8] &quot;I5CI33&quot; &quot;FB5L3N&quot; &quot;30J3CQ&quot; &quot;TEACA3&quot; &quot;7B9CA6&quot; &quot;0IIAEN&quot; &quot;BCJJKN&quot;
## [15] &quot;0X4W26&quot; &quot;JLFKV8&quot; &quot;DKIM6U&quot; &quot;GTLA8R&quot; &quot;SH3FB7&quot; &quot;1CZM30&quot; &quot;Q7U139&quot;
## [22] &quot;DCJJYS&quot; &quot;0SGJ12&quot; &quot;G25E3F&quot; &quot;QW2Z3R&quot; &quot;SHG3RB&quot; &quot;3SKITJ&quot; &quot;W0GUKI&quot;
## [29] &quot;T3QPW5&quot; &quot;6F9FB8&quot;
## 
## [[2]]
##  [1] &quot;3QHAFI&quot; &quot;1GF3GM&quot; &quot;87AQLF&quot; &quot;N5QBWD&quot; &quot;9P0DES&quot; &quot;3YJIMV&quot; &quot;WJXIH9&quot;
##  [8] &quot;W6MDVK&quot; &quot;83HQBN&quot; &quot;S7IWWA&quot; &quot;S222R3&quot; &quot;Q17CG3&quot; &quot;RJ4JPC&quot; &quot;HE0SCR&quot;
## [15] &quot;W5WIRP&quot; &quot;5EDLL7&quot; &quot;QQMBT1&quot; &quot;1SPLS8&quot; &quot;E5Q33K&quot; &quot;QRZK48&quot; &quot;PU7RSG&quot;
## [22] &quot;GIIEUD&quot; &quot;LVYYNY&quot; &quot;38K2SR&quot; &quot;MX4J7G&quot; &quot;6KWVRI&quot; &quot;WKY2SZ&quot; &quot;K3TNHP&quot;
## [29] &quot;ZPS15A&quot;
## 
## [[3]]
##  [1] &quot;8TV4MT&quot; &quot;CHK1ZX&quot; &quot;N79QXB&quot; &quot;0XTZQ1&quot; &quot;50D77I&quot; &quot;5W621W&quot; &quot;YLRNIK&quot;
##  [8] &quot;1SSCJC&quot; &quot;5BPBUI&quot; &quot;MYUMMX&quot; &quot;KX0RJ3&quot; &quot;7ZNY75&quot; &quot;S056D5&quot; &quot;D9P18Y&quot;
## [15] &quot;NK802Y&quot; &quot;AW400C&quot; &quot;XEC0M5&quot; &quot;TXZUKC&quot; &quot;N4NV8B&quot; &quot;C18V6I&quot; &quot;R5AYJK&quot;
## [22] &quot;M9PVG5&quot; &quot;AR5U44&quot; &quot;WTE53B&quot; &quot;B134XZ&quot; &quot;CLSVU6&quot; &quot;AP1YLW&quot; &quot;ZH3YG1&quot;
## [29] &quot;X694YR&quot; &quot;VWC5ZH&quot; &quot;F7I2ED&quot; &quot;3GECJJ&quot; &quot;B228Q6&quot;
## 
## [[4]]
##  [1] &quot;YDRD81&quot; &quot;MKY9TK&quot; &quot;01QRQ4&quot; &quot;LYSLPP&quot; &quot;QWKFBH&quot; &quot;Y6DB6L&quot; &quot;G58RGY&quot;
##  [8] &quot;7NE2UT&quot; &quot;CRPXY7&quot; &quot;DRXMW4&quot; &quot;MTCAIG&quot; &quot;Z904TJ&quot; &quot;5IAFMK&quot; &quot;DPXEQE&quot;
## [15] &quot;MH88T6&quot; &quot;EZ2F8A&quot; &quot;967Y3D&quot; &quot;AR17R5&quot; &quot;DH9WJQ&quot; &quot;BS3RLE&quot; &quot;99BMJW&quot;
## [22] &quot;CMMUKU&quot; &quot;D4B0RM&quot; &quot;DHNQ1W&quot; &quot;XFWVVX&quot; &quot;YTJ2UL&quot; &quot;321LLB&quot; &quot;PYPM1W&quot;
## [29] &quot;2F6J3U&quot; &quot;Y0TCYX&quot; &quot;ILVQVB&quot; &quot;MB6NYQ&quot; &quot;FL170P&quot; &quot;FG0SFA&quot; &quot;KZY6PD&quot;
## 
## [[5]]
##  [1] &quot;IZDV8K&quot; &quot;1VP3UC&quot; &quot;6X6BG9&quot; &quot;D33J06&quot; &quot;IH1KPA&quot; &quot;WK89I9&quot; &quot;PBAFJF&quot;
##  [8] &quot;PI4VHT&quot; &quot;QCENKM&quot; &quot;TQEMY6&quot; &quot;1FAZ0K&quot; &quot;J3F6PD&quot; &quot;1QVS67&quot; &quot;NN3GDQ&quot;
## [15] &quot;1KJ2MG&quot; &quot;8JUUJ9&quot; &quot;DI4AHD&quot; &quot;KEA4QG&quot; &quot;1CIRC9&quot; &quot;MFKT9C&quot; &quot;AZ3L0D&quot;
## [22] &quot;LS184H&quot; &quot;PJ72W1&quot; &quot;92UG4N&quot; &quot;SCFSBF&quot; &quot;I8ABC7&quot; &quot;GCBYDW&quot; &quot;Q8U9LB&quot;
## [29] &quot;XYRDKV&quot;
## 
## [[6]]
##  [1] &quot;A98D7P&quot; &quot;7RA57Q&quot; &quot;EX5K0S&quot; &quot;YFCIHJ&quot; &quot;WLMGS1&quot; &quot;2Z4YLY&quot; &quot;465ERA&quot;
##  [8] &quot;5KWNMZ&quot; &quot;H2J6UA&quot; &quot;S5H1GC&quot; &quot;F45799&quot; &quot;0HYZ23&quot; &quot;AFZKBS&quot; &quot;S3EBGZ&quot;
## [15] &quot;5ERY5Z&quot; &quot;MPIQ4N&quot; &quot;GAS52W&quot; &quot;QCA36T&quot; &quot;9MG040&quot; &quot;ZQXZYB&quot; &quot;WNEAS6&quot;
## [22] &quot;ZATMEE&quot; &quot;TYEWF1&quot; &quot;46ZHKN&quot; &quot;BKWE4D&quot; &quot;FJS7RQ&quot; &quot;ESUIAF&quot; &quot;CS23RV&quot;
## [29] &quot;5EDIEE&quot; &quot;3DTD2N&quot; &quot;RVHVTZ&quot; &quot;414N7M&quot; &quot;PVY432&quot; &quot;G8MCV7&quot;
## 
## [[7]]
## [1] NA</code></pre>
<p>We can identify and list the males in each group with the following code.</p>
<pre class="r"><code>sapply(haremGrp$group, function(ids) {ids[ids %in% trimmedPed$id[trimmedPed$sex == &quot;M&quot;]]})</code></pre>
<pre><code>## [[1]]
## [1] &quot;CHSCFG&quot;
## 
## [[2]]
## [1] &quot;3QHAFI&quot;
## 
## [[3]]
## [1] &quot;8TV4MT&quot;
## 
## [[4]]
## [1] &quot;YDRD81&quot;
## 
## [[5]]
## [1] &quot;IZDV8K&quot;
## 
## [[6]]
## [1] &quot;A98D7P&quot;
## 
## [[7]]
## logical(0)</code></pre>
<p>It is easy to notice that the male is listed first within each breeding group.</p>
<p>We can also see the number of animals and the sex ratios created in each group. Since these are harem groups the sex ratios are determined by the number of animals in the individual groups.</p>
<pre class="r"><code>lines &lt;- sapply(haremGrp$group, function(ids) {
  paste0(&quot;Count: &quot;, length(ids), &quot; Sex Ratio: &quot;, 
         round(calculateSexRatio(ids, trimmedPed), 2))})
for (line in lines) print(line)</code></pre>
<pre><code>## [1] &quot;Count: 30 Sex Ratio: 29&quot;
## [1] &quot;Count: 29 Sex Ratio: 28&quot;
## [1] &quot;Count: 33 Sex Ratio: 32&quot;
## [1] &quot;Count: 35 Sex Ratio: 34&quot;
## [1] &quot;Count: 29 Sex Ratio: 28&quot;
## [1] &quot;Count: 34 Sex Ratio: 33&quot;
## [1] &quot;Count: 1 Sex Ratio: Inf&quot;</code></pre>
<p>Examination of this table shows that of the 184 females 155 are included.</p>
</div>
<div id="section-controlling-sex-ratios" class="section level3">
<h3>Controlling Sex Ratios</h3>
<p>The following group assignments will be forming harem groups. This is done by setting <strong>harem</strong> to .</p>
<pre class="r"><code>sexRatioGrp &lt;- groupAddAssign(candidates = candidates,
                     kmat = trimmedGeneticValue[[&quot;kinship&quot;]],
                     ped = trimmedPed,
                     iter = 1000,
                     numGp = 6,
                     sexRatio = 9)
sexRatioGrp$group</code></pre>
<pre><code>## [[1]]
##  [1] &quot;1CIRC9&quot; &quot;ZDRSG0&quot; &quot;5EDLL7&quot; &quot;MYUMMX&quot; &quot;WKY2SZ&quot; &quot;AP1YLW&quot; &quot;CLSVU6&quot;
##  [8] &quot;MH88T6&quot; &quot;R5AYJK&quot; &quot;3SKITJ&quot; &quot;9P0DES&quot; &quot;AR5U44&quot; &quot;1KJ2MG&quot; &quot;0V4SAC&quot;
## [15] &quot;1SSCJC&quot; &quot;PJ72W1&quot; &quot;B134XZ&quot; &quot;D33J06&quot; &quot;Q17CG3&quot; &quot;DI4AHD&quot; &quot;S056D5&quot;
## [22] &quot;LS184H&quot; &quot;99BMJW&quot; &quot;LVYYNY&quot; &quot;GM371F&quot; &quot;K3TNHP&quot; &quot;WJXIH9&quot; &quot;S5H1GC&quot;
## [29] &quot;N79QXB&quot; &quot;1QVS67&quot; &quot;1VP3UC&quot; &quot;5KWNMZ&quot; &quot;38K2SR&quot; &quot;83HQBN&quot; &quot;IZDV8K&quot;
## 
## [[2]]
##  [1] &quot;WLMGS1&quot; &quot;DGZLV3&quot; &quot;F45799&quot; &quot;AFZKBS&quot; &quot;PI4VHT&quot; &quot;Y0TCYX&quot; &quot;WNEAS6&quot;
##  [8] &quot;GIIEUD&quot; &quot;FL170P&quot; &quot;5BPBUI&quot; &quot;CMMUKU&quot; &quot;92UG4N&quot; &quot;TQEMY6&quot; &quot;G2GYST&quot;
## [15] &quot;0XTZQ1&quot; &quot;GTLA8R&quot; &quot;XFWVVX&quot; &quot;N5QBWD&quot; &quot;TEACA3&quot; &quot;G8MCV7&quot; &quot;B228Q6&quot;
## [22] &quot;ZPS15A&quot; &quot;G25E3F&quot; &quot;PVY432&quot; &quot;KZM9RB&quot; &quot;I5CI33&quot; &quot;DKIM6U&quot; &quot;T3QPW5&quot;
## [29] &quot;1CZM30&quot; &quot;30J3CQ&quot; &quot;MKY9TK&quot; &quot;1FAZ0K&quot; &quot;MTCAIG&quot;
## 
## [[3]]
##  [1] &quot;B1WVCN&quot; &quot;XY2CK7&quot; &quot;BKWE4D&quot; &quot;KX0RJ3&quot; &quot;DHNQ1W&quot; &quot;01QRQ4&quot; &quot;QWKFBH&quot;
##  [8] &quot;2F6J3U&quot; &quot;46ZHKN&quot; &quot;5IAFMK&quot; &quot;BCJJKN&quot; &quot;ESUIAF&quot; &quot;CS23RV&quot; &quot;S63QDN&quot;
## [15] &quot;FG0SFA&quot; &quot;WTE53B&quot; &quot;AW400C&quot; &quot;IH1KPA&quot; &quot;0SGJ12&quot; &quot;C18V6I&quot; &quot;DH9WJQ&quot;
## [22] &quot;XYRDKV&quot; &quot;EZ2F8A&quot; &quot;WK89I9&quot; &quot;5KFB90&quot; &quot;G58RGY&quot; &quot;YFCIHJ&quot; &quot;YTJ2UL&quot;
## [29] &quot;465ERA&quot; &quot;DPXEQE&quot; &quot;MFKT9C&quot; &quot;PU7RSG&quot; &quot;KZY6PD&quot; &quot;GCBYDW&quot; &quot;Z25D52&quot;
## 
## [[4]]
##  [1] &quot;QRZK48&quot; &quot;SXSVEH&quot; &quot;FB5L3N&quot; &quot;9MG040&quot; &quot;NK802Y&quot; &quot;3YJIMV&quot; &quot;W5WIRP&quot;
##  [8] &quot;1SPLS8&quot; &quot;7B9CA6&quot; &quot;0HYZ23&quot; &quot;50D77I&quot; &quot;MB6NYQ&quot; &quot;ZQXZYB&quot; &quot;FFGPS4&quot;
## [15] &quot;13B1QL&quot; &quot;414N7M&quot; &quot;N4NV8B&quot; &quot;EX5K0S&quot; &quot;ILVQVB&quot; &quot;8JUUJ9&quot; &quot;TXZUKC&quot;
## [22] &quot;M9PVG5&quot; &quot;2Z4YLY&quot; &quot;6X6BG9&quot; &quot;3MMZD4&quot; &quot;W6MDVK&quot; &quot;Y6DB6L&quot; &quot;MPIQ4N&quot;
## [29] &quot;HE0SCR&quot;
## 
## [[5]]
##  [1] &quot;ZATMEE&quot; &quot;CHJ9D2&quot; &quot;87AQLF&quot; &quot;6F9FB8&quot; &quot;7RA57Q&quot; &quot;7NE2UT&quot; &quot;0IIAEN&quot;
##  [8] &quot;5ERY5Z&quot; &quot;NN3GDQ&quot; &quot;DCJJYS&quot; &quot;SHG3RB&quot; &quot;0X4W26&quot; &quot;967Y3D&quot; &quot;CFD12A&quot;
## [15] &quot;ZH3YG1&quot; &quot;F7I2ED&quot; &quot;RVHVTZ&quot; &quot;GAS52W&quot; &quot;SH3FB7&quot; &quot;CRPXY7&quot; &quot;AIHJ8Z&quot;
## [22] &quot;6KWVRI&quot; &quot;I8ABC7&quot; &quot;TYEWF1&quot; &quot;K7900I&quot; &quot;1GF3GM&quot; &quot;AR17R5&quot; &quot;EMV4P6&quot;
## [29] &quot;SCFSBF&quot; &quot;J3F6PD&quot; &quot;RJ4JPC&quot; &quot;S222R3&quot;
## 
## [[6]]
##  [1] &quot;Z904TJ&quot; &quot;YDRD81&quot; &quot;D9P18Y&quot; &quot;MX4J7G&quot; &quot;3GECJJ&quot; &quot;QCENKM&quot; &quot;AZ3L0D&quot;
##  [8] &quot;D4B0RM&quot; &quot;5W621W&quot; &quot;YLRNIK&quot; &quot;X694YR&quot; &quot;Q7U139&quot; &quot;KEA4QG&quot; &quot;AEP5EG&quot;
## [15] &quot;JLFKV8&quot; &quot;W0GUKI&quot; &quot;5EDIEE&quot; &quot;CHK1ZX&quot; &quot;Q8U9LB&quot; &quot;PBAFJF&quot; &quot;QQMBT1&quot;
## [22] &quot;321LLB&quot; &quot;WI38KZ&quot; &quot;VWC5ZH&quot; &quot;JSAP3H&quot; &quot;DRXMW4&quot; &quot;H2J6UA&quot; &quot;FJS7RQ&quot;
## [29] &quot;LYSLPP&quot; &quot;XEC0M5&quot; &quot;S3EBGZ&quot; &quot;S7IWWA&quot; &quot;E5Q33K&quot; &quot;BS3RLE&quot;
## 
## [[7]]
##  [1] &quot;HLQ9SY&quot; &quot;B2CKHA&quot; &quot;GDXWJ1&quot; &quot;TR5L57&quot; &quot;XC304E&quot; &quot;Z7NBA2&quot; &quot;1E8KD1&quot;
##  [8] &quot;5PW7WT&quot; &quot;A6A1M1&quot; &quot;BW10CL&quot; &quot;FTVE03&quot; &quot;IRFJ09&quot; &quot;KXHGRH&quot; &quot;LMJWTN&quot;
## [15] &quot;Q9LWGX&quot; &quot;R6HV9A&quot; &quot;RNQU14&quot; &quot;09LFE4&quot; &quot;3DTD2N&quot; &quot;3QHAFI&quot; &quot;55BPSE&quot;
## [22] &quot;5XVTVH&quot; &quot;72LYDE&quot; &quot;8IG767&quot; &quot;9FRCIE&quot; &quot;ER464J&quot; &quot;FG6L7S&quot; &quot;NHWTJ9&quot;
## [29] &quot;P7RBPI&quot; &quot;TBCE78&quot; &quot;YI16QD&quot; &quot;2F1IV1&quot; &quot;4LHK19&quot; &quot;59NYZE&quot; &quot;5IYDXN&quot;
## [36] &quot;6KLWVC&quot; &quot;80F2MI&quot; &quot;8TV4MT&quot; &quot;A98D7P&quot; &quot;AZ4D19&quot; &quot;BTTHAJ&quot; &quot;CHSCFG&quot;
## [43] &quot;EEGLWY&quot; &quot;FX9E4X&quot; &quot;G91ZM6&quot; &quot;J1R2EW&quot; &quot;LDND6J&quot; &quot;LN1DLY&quot; &quot;MQT080&quot;
## [50] &quot;NSIC4I&quot; &quot;PHB6TE&quot; &quot;PYPM1W&quot; &quot;QRWYQZ&quot; &quot;QW2Z3R&quot; &quot;RY1AZM&quot; &quot;WHQLH5&quot;
## [57] &quot;WQUN84&quot; &quot;XL658N&quot; &quot;XX0GYV&quot; &quot;YHHVC7&quot; &quot;YP910X&quot; &quot;0X1RZ9&quot; &quot;3YHBC1&quot;
## [64] &quot;55VDSQ&quot; &quot;653J82&quot; &quot;6MEP2C&quot; &quot;76DIT4&quot; &quot;7ZNY75&quot; &quot;80KACX&quot; &quot;B2YJJP&quot;
## [71] &quot;E3JP0C&quot; &quot;FLIZQI&quot; &quot;MEUZ85&quot; &quot;PA9F3J&quot; &quot;QCA36T&quot; &quot;T38W6H&quot; &quot;TJN1AD&quot;
## [78] &quot;WNKKW3&quot; &quot;XZH41H&quot; &quot;3P9BX6&quot; &quot;7D09WH&quot; &quot;ZW2X4N&quot;</code></pre>
<p>Again we can identify and list the males in each group with the following code.</p>
<pre class="r"><code>sapply(sexRatioGrp$group, function(ids) {ids[ids %in% trimmedPed$id[trimmedPed$sex == &quot;M&quot;]]})</code></pre>
<pre><code>## [[1]]
## [1] &quot;ZDRSG0&quot; &quot;0V4SAC&quot; &quot;GM371F&quot; &quot;IZDV8K&quot;
## 
## [[2]]
## [1] &quot;DGZLV3&quot; &quot;G2GYST&quot; &quot;KZM9RB&quot;
## 
## [[3]]
## [1] &quot;XY2CK7&quot; &quot;S63QDN&quot; &quot;5KFB90&quot; &quot;Z25D52&quot;
## 
## [[4]]
## [1] &quot;SXSVEH&quot; &quot;FFGPS4&quot; &quot;3MMZD4&quot;
## 
## [[5]]
## [1] &quot;CHJ9D2&quot; &quot;CFD12A&quot; &quot;K7900I&quot;
## 
## [[6]]
## [1] &quot;YDRD81&quot; &quot;AEP5EG&quot; &quot;JSAP3H&quot;
## 
## [[7]]
##  [1] &quot;HLQ9SY&quot; &quot;B2CKHA&quot; &quot;GDXWJ1&quot; &quot;TR5L57&quot; &quot;XC304E&quot; &quot;Z7NBA2&quot; &quot;1E8KD1&quot;
##  [8] &quot;5PW7WT&quot; &quot;A6A1M1&quot; &quot;BW10CL&quot; &quot;FTVE03&quot; &quot;IRFJ09&quot; &quot;KXHGRH&quot; &quot;LMJWTN&quot;
## [15] &quot;Q9LWGX&quot; &quot;R6HV9A&quot; &quot;RNQU14&quot; &quot;09LFE4&quot; &quot;3QHAFI&quot; &quot;55BPSE&quot; &quot;5XVTVH&quot;
## [22] &quot;8IG767&quot; &quot;9FRCIE&quot; &quot;ER464J&quot; &quot;FG6L7S&quot; &quot;NHWTJ9&quot; &quot;P7RBPI&quot; &quot;TBCE78&quot;
## [29] &quot;YI16QD&quot; &quot;2F1IV1&quot; &quot;4LHK19&quot; &quot;59NYZE&quot; &quot;5IYDXN&quot; &quot;6KLWVC&quot; &quot;80F2MI&quot;
## [36] &quot;8TV4MT&quot; &quot;A98D7P&quot; &quot;AZ4D19&quot; &quot;BTTHAJ&quot; &quot;CHSCFG&quot; &quot;EEGLWY&quot; &quot;FX9E4X&quot;
## [43] &quot;G91ZM6&quot; &quot;J1R2EW&quot; &quot;LDND6J&quot; &quot;LN1DLY&quot; &quot;MQT080&quot; &quot;NSIC4I&quot; &quot;PHB6TE&quot;
## [50] &quot;QRWYQZ&quot; &quot;RY1AZM&quot; &quot;WHQLH5&quot; &quot;WQUN84&quot; &quot;XL658N&quot; &quot;XX0GYV&quot; &quot;YHHVC7&quot;
## [57] &quot;YP910X&quot; &quot;0X1RZ9&quot; &quot;3YHBC1&quot; &quot;55VDSQ&quot; &quot;653J82&quot; &quot;6MEP2C&quot; &quot;76DIT4&quot;
## [64] &quot;80KACX&quot; &quot;B2YJJP&quot; &quot;E3JP0C&quot; &quot;FLIZQI&quot; &quot;MEUZ85&quot; &quot;PA9F3J&quot; &quot;T38W6H&quot;
## [71] &quot;TJN1AD&quot; &quot;WNKKW3&quot; &quot;XZH41H&quot; &quot;3P9BX6&quot; &quot;7D09WH&quot; &quot;ZW2X4N&quot;</code></pre>
<p>We can also see the number of animals and the sex ratios created in each group.</p>
<pre class="r"><code>lines &lt;- sapply(sexRatioGrp$group, function(ids) {
  paste0(&quot;Count: &quot;, length(ids), &quot; Sex Ratio: &quot;, 
         round(calculateSexRatio(ids, trimmedPed), 2))})
for (line in lines) print(line)</code></pre>
<pre><code>## [1] &quot;Count: 35 Sex Ratio: 7.75&quot;
## [1] &quot;Count: 33 Sex Ratio: 10&quot;
## [1] &quot;Count: 35 Sex Ratio: 7.75&quot;
## [1] &quot;Count: 29 Sex Ratio: 8.67&quot;
## [1] &quot;Count: 32 Sex Ratio: 9.67&quot;
## [1] &quot;Count: 34 Sex Ratio: 10.33&quot;
## [1] &quot;Count: 82 Sex Ratio: 0.08&quot;</code></pre>
<p>Examination of this table shows that of the 184 females 239 are included.</p>
</div>
</div>
<div id="section-pedigree-errors" class="section level2">
<h2>Pedigree Errors</h2>
<p>As stated earlier you can see which types of errors are detected by <em>qcStudbook</em> by looking at names returned by <em>getEmptyErrorLst()</em> as shown below.</p>
<pre class="r"><code>names(getEmptyErrorLst())</code></pre>
<pre><code>## [1] &quot;failedDatabaseConnection&quot; &quot;missingColumns&quot;          
## [3] &quot;invalidDateRows&quot;          &quot;suspiciousParents&quot;       
## [5] &quot;femaleSires&quot;              &quot;maleDams&quot;                
## [7] &quot;sireAndDam&quot;               &quot;duplicateIds&quot;            
## [9] &quot;changedCols&quot;</code></pre>
<p>Each is defined below.</p>
<table>
<thead>
<tr class="header">
<th align="left">Error</th>
<th align="left">Definition</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">failedDatabaseConnection</td>
<td align="left">Database connection failed: configuration or permissions are invalid</td>
</tr>
<tr class="even">
<td align="left">missingColumns</td>
<td align="left">Columns that must be within the pedigree file are missing.</td>
</tr>
<tr class="odd">
<td align="left">invalidDateRows</td>
<td align="left">Values, which are supposed to be dates, cannot be interpreted as a date.</td>
</tr>
<tr class="even">
<td align="left">suspiciousParents</td>
<td align="left">Parents were too young on the date of birth of to have been the parent.</td>
</tr>
<tr class="odd">
<td align="left">femaleSires</td>
<td align="left">Individuals listed as female or hemaphroditic and as a sire.</td>
</tr>
<tr class="even">
<td align="left">maleDams</td>
<td align="left">Individuals are listed as male and as a dam.</td>
</tr>
<tr class="odd">
<td align="left">sireAndDam</td>
<td align="left">Individuals who are listed as both a sire and a dam.</td>
</tr>
<tr class="even">
<td align="left">duplicateIds</td>
<td align="left">IDs listed more than once.</td>
</tr>
<tr class="odd">
<td align="left">changedCols</td>
<td align="left">Columns that have been changed to conform to internal naming conventions and what they were changed to.</td>
</tr>
</tbody>
</table>
<p>We are going to use the small imaginary pedigree listed below that has multiple errors to discuss pedigree error detection and reporting. First note the birthdates of ego_id <em>o4</em> (2006-04-13) and the purported sire <em>s2</em> (2006-06-19). Note also the proported birthdate of the <em>d2</em> and the birthdates of her offspring. Obviously dates or IDs may be incorrect.</p>
<p>This is the pedigree. <em>(We will discuss the column names shortly.)</em></p>
<pre class="r"><code>knitr::kable(nprcmanager::pedOne)</code></pre>
<table>
<thead>
<tr class="header">
<th align="left">ego_id</th>
<th align="left">si re</th>
<th align="left">dam_id</th>
<th align="left">sex</th>
<th align="left">birth_date</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">s1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">F</td>
<td align="left">2000-07-18</td>
</tr>
<tr class="even">
<td align="left">d1</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">M</td>
<td align="left">2003-04-13</td>
</tr>
<tr class="odd">
<td align="left">s2</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">M</td>
<td align="left">2006-06-19</td>
</tr>
<tr class="even">
<td align="left">d2</td>
<td align="left">NA</td>
<td align="left">NA</td>
<td align="left">F</td>
<td align="left">2015-09-16</td>
</tr>
<tr class="odd">
<td align="left">o1</td>
<td align="left">s1</td>
<td align="left">d1</td>
<td align="left">F</td>
<td align="left">2015-02-04</td>
</tr>
<tr class="even">
<td align="left">o2</td>
<td align="left">s1</td>
<td align="left">d2</td>
<td align="left">F</td>
<td align="left">2009-03-17</td>
</tr>
<tr class="odd">
<td align="left">o3</td>
<td align="left">s2</td>
<td align="left">d2</td>
<td align="left">F</td>
<td align="left">2012-04-11</td>
</tr>
<tr class="even">
<td align="left">o4</td>
<td align="left">s2</td>
<td align="left">d2</td>
<td align="left">M</td>
<td align="left">2006-04-13</td>
</tr>
</tbody>
</table>
<p>If we try to convert this pedigree file into the standardized studbook format, we are going to get an error message and the creation of a file in the user’s home directory that lists the records that have generated the errors.</p>
<pre class="r"><code>pedOne &lt;- nprcmanager::pedOne # put it in the local environment
ped &lt;- qcStudbook(pedOne, minParentAge = 0)</code></pre>
<pre><code>## Error in qcStudbook(pedOne, minParentAge = 0): Parents with low age at birth of offspring are listed in ~/lowParentAge.csv.</code></pre>
<p>The contents of <em>lowParentAge.csv</em> is shown below.</p>
<table>
<thead>
<tr class="header">
<th align="left">dam</th>
<th align="left">sire</th>
<th align="left">id</th>
<th align="left">sex</th>
<th align="left">birth</th>
<th align="left">recordStatus</th>
<th align="left">exit</th>
<th align="left">sireBirth</th>
<th align="left">damBirth</th>
<th align="right">sireAge</th>
<th align="right">damAge</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">d2</td>
<td align="left">s1</td>
<td align="left">o2</td>
<td align="left">F</td>
<td align="left">2009-03-17</td>
<td align="left">original</td>
<td align="left">NA</td>
<td align="left">2000-07-18</td>
<td align="left">2015-09-16</td>
<td align="right">8.67</td>
<td align="right">-6.50</td>
</tr>
<tr class="even">
<td align="left">d2</td>
<td align="left">s2</td>
<td align="left">o3</td>
<td align="left">F</td>
<td align="left">2012-04-11</td>
<td align="left">original</td>
<td align="left">NA</td>
<td align="left">2006-06-19</td>
<td align="left">2015-09-16</td>
<td align="right">5.82</td>
<td align="right">-3.43</td>
</tr>
<tr class="odd">
<td align="left">d2</td>
<td align="left">s2</td>
<td align="left">o4</td>
<td align="left">M</td>
<td align="left">2006-04-13</td>
<td align="left">original</td>
<td align="left">NA</td>
<td align="left">2006-06-19</td>
<td align="left">2015-09-16</td>
<td align="right">-0.18</td>
<td align="right">-9.43</td>
</tr>
</tbody>
</table>
<p>Examination of the ages of the parents reveals the issues being reported.</p>
<p>We can remove the errors by changing the birthdates of <em>o4</em> from 2006-04-13 to 2015-09-16 and of <em>d2</em> from 2015-09-16 to 2006-04-13.</p>
<pre class="r"><code>pedOne$birth_date[pedOne$ego_id == &quot;o4&quot;] &lt;- as.Date(&quot;2015-09-16&quot;)
pedOne$birth_date[pedOne$ego_id == &quot;d2&quot;] &lt;- as.Date(&quot;2006-04-13&quot;)</code></pre>
<p>Note the changes made to the column names between the original <strong>pedOne</strong> pedigree and the pedigree (<strong>ped</strong>) we get from <strong>qcStudbook</strong>. We have chosen to limit the displayed pedigree by selecting the <em>ego_id</em>’s and <em>id</em>’s in <strong>pedOne</strong> and <strong>ped</strong> respectively.</p>
<pre class="r"><code>ped &lt;- qcStudbook(pedOne, minParentAge = 0)
ped[ped$id %in% c(&quot;s2&quot;, &quot;d2&quot;, &quot;o3&quot;, &quot;o4&quot;), ]</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["id"],"name":[1],"type":["chr"],"align":["left"]},{"label":["sire"],"name":[2],"type":["chr"],"align":["left"]},{"label":["dam"],"name":[3],"type":["chr"],"align":["left"]},{"label":["sex"],"name":[4],"type":["fctr"],"align":["left"]},{"label":["gen"],"name":[5],"type":["dbl"],"align":["right"]},{"label":["birth"],"name":[6],"type":["date"],"align":["right"]},{"label":["exit"],"name":[7],"type":["date"],"align":["right"]},{"label":["age"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["recordStatus"],"name":[9],"type":["chr"],"align":["left"]}],"data":[{"1":"d2","2":"NA","3":"NA","4":"F","5":"0","6":"2006-04-13","7":"<NA>","8":"13.6","9":"original","_rn_":"2"},{"1":"s2","2":"NA","3":"NA","4":"M","5":"0","6":"2006-06-19","7":"<NA>","8":"13.4","9":"original","_rn_":"4"},{"1":"o3","2":"s2","3":"d2","4":"F","5":"1","6":"2012-04-11","7":"<NA>","8":"7.6","9":"original","_rn_":"7"},{"1":"o4","2":"s2","3":"d2","4":"M","5":"1","6":"2015-09-16","7":"<NA>","8":"4.1","9":"original","_rn_":"8"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>However, the preferred method of creating the standardized studbook format with <strong>qcStudbook</strong> is to examine all errors found and correcting them before proceeding. This is done by explicitely requesting that all aspects inconsistent with the studbook format be identified by setting <em>reportChanges</em> and <em>reportErrors</em> to .</p>
<pre class="r"><code>errorList &lt;- qcStudbook(pedOne, minParentAge = 0, reportChanges = TRUE, 
                        reportErrors = TRUE)
summary(errorList)</code></pre>
<pre><code>## Error: The animal listed as a sire and also listed as a female is: s1.
## Error: The animal listed as a dam and also listed as a male is: d1.
## Change: The column where space was removed is: si re to sire.
## Change: The columns where underscore was removed are: ego_id, dam_id, and birth_date to egoid, damid, and birthdate.
## Change: The column changed from: egoid to id.
## Change: The column changed from: damid to dam.
## Change: The column changed from: birthdate to birth.
## 
## Please check and correct the pedigree file.
## </code></pre>
<p>We will discuss each of these newly identified errors in a moment, however, let’s look at shortening this report, because often you will not be interested in the more trivial changes to the column names made by <strong>qcStudbook</strong> and in those cases you simply choose not to report changes to the column names as is shown here by setting <em>reportChanges</em> to . For this illustration, we are going to bring back the original copy of <strong>pedOne</strong> to see how the suspicious parents are reported by the <strong>summmary</strong> function.</p>
<pre class="r"><code>pedOne &lt;- nprcmanager::pedOne
errorList &lt;- qcStudbook(pedOne, minParentAge = 0, reportChanges = FALSE, 
                        reportErrors = TRUE)
options(width = 90)
summary(errorList)</code></pre>
<pre><code>## Error: The animal listed as a sire and also listed as a female is: s1.
## Error: The animal listed as a dam and also listed as a male is: d1.
## 
## Please check and correct the pedigree file.
##  
## Animal records where parent records are suspicous because of dates.
## One or more parents appear too young at time of birth.
##   dam sire id sex      birth recordStatus exit  sireBirth   damBirth sireAge damAge
## 2  d2   s1 o2   F 2009-03-17     original &lt;NA&gt; 2000-07-18 2015-09-16    8.67   -6.5
## 3  d2   s2 o3   F 2012-04-11     original &lt;NA&gt; 2006-06-19 2015-09-16    5.82   -3.4
## 4  d2   s2 o4   M 2006-04-13     original &lt;NA&gt; 2006-06-19 2015-09-16   -0.18   -9.4</code></pre>
<p>The first two errors mentioned are of particular interest. Currently <strong>qcStudbook</strong> automatically changes the sex of dams to <em>F</em> (female) and sires to <em>M</em> (male) when <strong>reportErrors</strong> is set to .</p>
</div>
<div id="section-genetic-loops" class="section level2">
<h2>Genetic Loops</h2>
<p>This feature is not supported within the Shiny application and is not fully implemented.</p>
<p>To use the <strong>findLoops</strong> function run the following code and select a pedigree as your input file that has a loop in it. We are continuing to use the example pedigree that comes with the software <em>Example_Pedigree.csv</em>.</p>
<pre class="r"><code>exampleTree &lt;- createPedTree(breederPed)
exampleLoops &lt;- findLoops(exampleTree)</code></pre>
<p>You can count how many loops you have with the following code.</p>
<pre class="r"><code>length(exampleLoops[exampleLoops == TRUE])</code></pre>
<pre><code>## [1] 145</code></pre>
<pre class="r"><code>nLoops &lt;- countLoops(exampleLoops, exampleTree)
sum(unlist(nLoops[nLoops &gt; 0]))</code></pre>
<pre><code>## [1] 258</code></pre>
<p>You can list the first 10 sets of ids, sires and dams in loops with the following line of code:</p>
<pre class="r"><code>examplePedigree[exampleLoops == TRUE, c(&quot;id&quot;, &quot;sire&quot;, &quot;dam&quot;)][1:10, ]</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["id"],"name":[1],"type":["chr"],"align":["left"]},{"label":["sire"],"name":[2],"type":["chr"],"align":["left"]},{"label":["dam"],"name":[3],"type":["chr"],"align":["left"]}],"data":[{"1":"V49H3Y","2":"UFI88T","3":"9T7Y2Z","_rn_":"2519"},{"1":"61FUGE","2":"UDQ5WC","3":"GL88CF","_rn_":"2572"},{"1":"LWJ3A5","2":"KZM9RB","3":"GCBYDW","_rn_":"2695"},{"1":"RNQU14","2":"H2RDE2","3":"DKIM6U","_rn_":"2722"},{"1":"L9M1DC","2":"3PU50K","3":"WFQENR","_rn_":"2752"},{"1":"Q8U9LB","2":"3PU50K","3":"CLSVU6","_rn_":"2755"},{"1":"FVJ14K","2":"UXC40T","3":"L5VC2M","_rn_":"2905"},{"1":"531HAC","2":"UMV4BE","3":"5DIPZN","_rn_":"2922"},{"1":"85ESBB","2":"UQFY9C","3":"Q2RK1E","_rn_":"2924"},{"1":"0VLW56","2":"6KPKH7","3":"MMEHXV","_rn_":"2941"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<pre class="r"><code>elapsed_time &lt;- get_elapsed_time_str(start_time)</code></pre>
<p>The current date and time is 2019-11-09 21:23:21. The processing time for this document was 2 minutes and 24 seconds..</p>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Catalina 10.15
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] testthat_2.2.1      futile.logger_1.4.3 DT_0.9              shinyBS_0.61       
##  [5] png_0.1-7           knitr_1.25          ggplot2_3.2.1       nprcmanager_0.5.36 
##  [9] stringi_1.4.3       learnr_0.10.0       shiny_1.4.0        
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.2           lubridate_1.7.4      lattice_0.20-38      gtools_3.8.1        
##  [5] assertthat_0.2.1     rprojroot_1.3-2      digest_0.6.21        mime_0.7            
##  [9] R6_2.4.0             cellranger_1.1.0     futile.options_1.0.1 backports_1.1.5     
## [13] WriteXLS_5.0.0       evaluate_0.14        highr_0.8            httr_1.4.1          
## [17] pillar_1.4.2         gplots_3.0.1.1       rlang_0.4.0          lazyeval_0.2.2      
## [21] readxl_1.3.1         rstudioapi_0.10      gdata_2.18.0         Matrix_1.2-17       
## [25] checkmate_1.9.4      rmarkdown_1.16       labeling_0.3         stringr_1.4.0       
## [29] htmlwidgets_1.5.1    munsell_0.5.0        anytime_0.3.6        compiler_3.6.1      
## [33] httpuv_1.5.2         xfun_0.10            pkgconfig_2.0.3      htmltools_0.4.0     
## [37] tidyselect_0.2.5     tibble_2.1.3         htmlTable_1.13.2     codetools_0.2-16    
## [41] crayon_1.3.4         dplyr_0.8.3          withr_2.1.2          later_1.0.0         
## [45] bitops_1.0-6         grid_3.6.1           jsonlite_1.6         xtable_1.8-4        
## [49] gtable_0.3.0         magrittr_1.5         formatR_1.7          scales_1.0.0        
## [53] KernSmooth_2.23-16   promises_1.1.0       lambda.r_1.2.4       RColorBrewer_1.1-2  
## [57] tools_3.6.1          glue_1.3.1           Rlabkey_2.3.2        purrr_0.3.3         
## [61] markdown_1.1         plotrix_3.7-6        fastmap_1.0.1        yaml_2.2.0          
## [65] colorspace_1.4-1     caTools_1.17.1.2</code></pre>

<script type="application/shiny-prerendered" data-context="server-start">
library(learnr)
knitr::opts_chunk$set(echo = TRUE,include = TRUE, results = "markup", cache = TRUE)
start_time <- proc.time()
</script>
 
<script type="application/shiny-prerendered" data-context="server">
learnr:::register_http_handlers(session, metadata = NULL)
</script>
 <!--html_preserve-->
<script type="application/shiny-prerendered" data-context="dependencies">
{"type":"list","attributes":{},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.16"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootstrap"]},{"type":"character","attributes":{},"value":["3.3.5"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/bootstrap"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["viewport"]}},"value":[{"type":"character","attributes":{},"value":["width=device-width, initial-scale=1"]}]},{"type":"character","attributes":{},"value":["js/bootstrap.min.js","shim/html5shiv.min.js","shim/respond.min.js"]},{"type":"character","attributes":{},"value":["css/cerulean.min.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.16"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["pagedtable"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/pagedtable-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["js/pagedtable.js"]},{"type":"character","attributes":{},"value":["css/pagedtable.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.16"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["textmate.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.16"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-autocompletion"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-autocompletion.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-diagnostics"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-diagnostics.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-format"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmarkdown/templates/tutorial/resources"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-format.js"]},{"type":"character","attributes":{},"value":["tutorial-format.css","rstudio-theme.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.16"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["navigation"]},{"type":"character","attributes":{},"value":["1.1"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/navigation-1.1"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tabsets.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.16"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["highlightjs"]},{"type":"character","attributes":{},"value":["9.12.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/highlightjs"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["highlight.js"]},{"type":"character","attributes":{},"value":["default.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.16"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["jquery"]},{"type":"character","attributes":{},"value":["1.11.3"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/jquery"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["jquery.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.16"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["font-awesome"]},{"type":"character","attributes":{},"value":["5.1.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["rmd/h/fontawesome"]}]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["css/all.css","css/v4-shims.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["rmarkdown"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["1.16"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["bootbox"]},{"type":"character","attributes":{},"value":["4.4.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/bootbox"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["bootbox.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["idb-keyvalue"]},{"type":"character","attributes":{},"value":["3.2.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/idb-keyval"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["idb-keyval-iife-compat.min.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[false]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial.js"]},{"type":"character","attributes":{},"value":["tutorial.css"]},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-autocompletion"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-autocompletion.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["name","version","src","meta","script","stylesheet","head","attachment","package","all_files","pkgVersion"]},"class":{"type":"character","attributes":{},"value":["html_dependency"]}},"value":[{"type":"character","attributes":{},"value":["tutorial-diagnostics"]},{"type":"character","attributes":{},"value":["0.10.0"]},{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["file"]}},"value":[{"type":"character","attributes":{},"value":["lib/tutorial"]}]},{"type":"NULL"},{"type":"character","attributes":{},"value":["tutorial-diagnostics.js"]},{"type":"NULL"},{"type":"NULL"},{"type":"NULL"},{"type":"character","attributes":{},"value":["learnr"]},{"type":"logical","attributes":{},"value":[true]},{"type":"character","attributes":{},"value":["0.10.0"]}]}]}
</script>
<!--/html_preserve-->
<!--html_preserve-->
<script type="application/shiny-prerendered" data-context="execution_dependencies">
{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages"]}},"value":[{"type":"list","attributes":{"names":{"type":"character","attributes":{},"value":["packages","version"]},"class":{"type":"character","attributes":{},"value":["data.frame"]},"row.names":{"type":"integer","attributes":{},"value":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84]}},"value":[{"type":"character","attributes":{},"value":["anytime","assertthat","backports","base","bitops","caTools","cellranger","checkmate","codetools","colorspace","compiler","crayon","datasets","digest","dplyr","DT","evaluate","fastmap","formatR","futile.logger","futile.options","gdata","ggplot2","glue","gplots","graphics","grDevices","grid","gtable","gtools","highr","htmlTable","htmltools","htmlwidgets","httpuv","httr","jsonlite","KernSmooth","knitr","labeling","lambda.r","later","lattice","lazyeval","learnr","lubridate","magrittr","markdown","Matrix","methods","mime","munsell","nprcmanager","pillar","pkgconfig","plotrix","png","promises","purrr","R6","RColorBrewer","Rcpp","readxl","Rlabkey","rlang","rmarkdown","rprojroot","rstudioapi","scales","shiny","shinyBS","stats","stringi","stringr","testthat","tibble","tidyselect","tools","utils","withr","WriteXLS","xfun","xtable","yaml"]},{"type":"character","attributes":{},"value":["0.3.6","0.2.1","1.1.5","3.6.1","1.0-6","1.17.1.2","1.1.0","1.9.4","0.2-16","1.4-1","3.6.1","1.3.4","3.6.1","0.6.21","0.8.3","0.9","0.14","1.0.1","1.7","1.4.3","1.0.1","2.18.0","3.2.1","1.3.1","3.0.1.1","3.6.1","3.6.1","3.6.1","0.3.0","3.8.1","0.8","1.13.2","0.4.0","1.5.1","1.5.2","1.4.1","1.6","2.23-16","1.25","0.3","1.2.4","1.0.0","0.20-38","0.2.2","0.10.0","1.7.4","1.5","1.1","1.2-17","3.6.1","0.7","0.5.0","0.5.36","1.4.2","2.0.3","3.7-6","0.1-7","1.1.0","0.3.3","2.4.0","1.1-2","1.0.2","1.3.1","2.3.2","0.4.0","1.16","1.3-2","0.10","1.0.0","1.4.0","0.61","3.6.1","1.4.3","1.4.0","2.2.1","2.1.3","0.2.5","3.6.1","3.6.1","2.1.2","5.0.0","0.10","1.8-4","2.2.0"]}]}]}
</script>
<!--/html_preserve-->
</div>
<div class="footnotes">
<hr />
<ol>
<li id="section-fn1"><p>Setting the <em>minParentAge</em> to 3.5 and above will cause an error along with the creation of a file <em>~/lowParentAge.csv</em> that will list the parents with low age at the birth of an offspring.<a href="#section-fnref1" class="footnote-back">↩</a></p></li>
<li id="section-fn2"><p>The <em>population</em> column is created and added to the pedigree object if it does not already exist.<a href="#section-fnref2" class="footnote-back">↩</a></p></li>
<li id="section-fn3"><p>All animals within the colony have a known birth date.<a href="#section-fnref3" class="footnote-back">↩</a></p></li>
</ol>
</div>

</div> <!-- topics -->

<div class="topicsContainer">
<div class="topicsPositioner">
<div class="band">
<div class="bandContent topicsListContainer">

<!-- begin doc-metadata -->
<div id="doc-metadata">
<h2 class="title toc-ignore" style="display:none;">Interactive Use of nprcmanager</h2>
</div>
<!-- end doc-metadata -->

</div> <!-- bandContent.topicsListContainer -->
</div> <!-- band -->
</div> <!-- topicsPositioner -->
</div> <!-- topicsContainer -->


</div> <!-- bandContent page -->
</div> <!-- pageContent band -->




<script>
// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>


</body>

</html>
