<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Implements Genetic Tools for Colony Management • nprcmanager</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">nprcmanager</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="/index.html">
    <span class="fa fa-home fa-lg"></span>

  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Development_Plans.html">Development Plans</a>
    </li>
    <li>
      <a href="articles/Notebook.html">Genetic Management Tools</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->


      </header><div class="row">
  <div class="contents col-md-9">

<div class="page-header"><h1>Genetic Tools for Colony Management</h1></div>
<h3>Genetic Value Analysis</h3>
<p>
    The genetic value analysis proceeds according to the following steps:
</p>
<ol type="1">
<li style="padding-bottom:15px">
        First, the average of all pairwise <b>kinship coefficients</b> are
        calculated for each living animal in the breeding colony relative
        to every other living animal in the colony according, to the algorithm
        of Lange (2002). The grand mean kinship in the breeding colony is then
        calculated from the distribution of individual mean kinships, along with
        other summary statistics describing this distribution.
    </li>
    <li style="padding-bottom:15px">
        <p>
        Second, measures of individual mean kinship for each animal are then
        calculated in terms of standard deviates from this grand mean kinship,
        as <em>z = (x-μ)/σ</em>, i.e., "<b>z-scores</b>", in which x is the estimate of
        individual mean kinship, μ is the grand mean kinship, and σ is the
        standard deviation of the distribution of all individual mean kinship
        values of living animals in the colony.
        </p>
        <p>
        These z-scores indicate the extent and direction to which each animal's
        individual mean kinship deviates from the grand mean kinship in the colony.
        An animal having a positive z-score has individual mean kinship that lies
        above the colony grand mean kinship, suggesting that this animal's genome
        is over-represented in the colony, making the animal of lesser genetic value
        than others with negative z-scores.  Conversely, an animal having a negative
        z-score has individual mean kinship that lies below the grand mean kinship,
        suggesting that this animal's genome is underrepresented in the colony, making
        the animal of greater genetic value than animals with positive z-scores.
        </p>
    </li>
    <li style="padding-bottom:15px">
        Third, <b>genome uniqueness</b> values are calculated using a gene-drop simulation
        according to MacCluer et al. (1986) and Ballou &amp; Lacy (1995), by assigning unique
        alleles to all pedigree founders, and simulating their segregation throughout the
        pedigree according to Mendelian rules.  Genome uniqueness is a measure of the
        probability that an animal possesses founder alleles that are present in at most
        <em>x</em> other animals (usually 0-3), and thus are rare and at risk of being lost from
        the population.  A range of 2 to 100,000 simulations may be selected.  A minimum
        of 1,000 simulations is recommended.  A genome uniqueness threshold value between
        0-3 should also be selected, as desired.
    </li>
    <li style="padding-bottom:15px">
        Following the calculation of mean kinship, related z-scores, and genome uniqueness
        for all animals, animals are ranked from the highest genetic value (ranking = 1)
        to the lowest, according to the following rules:
        <ol type="i">
<li style="padding-top:5px; padding-bottom:5px">
                Imported animals without offspring are automatically ranked highest in
                genetic value, followed by
            </li>
            <li style="padding-bottom:5px">
                All animals with greater than 10% genome uniqueness, ordered first by
                genome uniqueness from largest to smallest values, and by z-score within
                the same genome uniqueness from smallest to largest values; followed by
            </li>
            <li style="padding-bottom:5px">
                All remaining animals (those with = 10% genome uniqueness), ranked by
                z-score from smallest to largest values; followed by
            </li>
            <li>
                A threshold drawn at those animals with genome uniqueness = 10% and
                z-scores = 0.25.  Animals falling below this threshold are indicated by
                highlighting in the results.  Animals ranking above this threshold are
                considered to be of high genetic value, while animals that rank below this
                threshold are not recommended for breeding, but may be suitable for research
                assignment or sale.
            </li>
        </ol>
</li>
</ol>
<p><br></p>
<h3>Breeding Group Formation</h3> <ul style="list-style-type:disc">
<li style="padding-bottom: 15px"> In large colonies, most or all genetic management may take place primarily during development of the housing/breeding group. A genetic value analysis may be used as the basis for creating potential new breeding groups that will maximize genetic diversity within the colony as a whole. </li> <li style="padding-bottom: 15px"> When using the results of a genetic value analysis to develop new breeding groups, it is important to recognize that animals of high genetic value may still be closely related to other high-value animals, and that genetic diversity in the colony can be greatly reduced if relatives are allowed to produce viable offspring. </li> <li style="padding-bottom: 15px"> Following a genetic value analysis of all breeding colony animals, only animals lying above the threshold will be considered for potential new breeding groups. In this analysis, potential new breeding groups are formed based on relationships that must be excluded, the threshold value of within-group relatedness that may be tolerated, relationships that can be safely ignored, and any additional specific group criteria. For example, relatedness greater than second cousin between breeding males, and between breeding males and females should be of concern; however, relatedness between breeding adults and animals ≤ 1 year of age, or between females of any age might be ignored, depending on the social group structure desired. </li> <li style="padding-bottom: 15px"> This algorithm will assign randomly-selected animals to hypothetical groups, with the constraint that they cannot be related at the specified threshold level or greater to animals already assigned to the group. This process is repeated 10,000 times to thoroughly sample the collection of possible groups that can be formed from the high-value set of candidate animals being considered. Upon completion, this algorithm will return the best set of groups, defined as the set with the largest minimum group size. </li> </ul>
</div>

  <div class="col-md-3" id="sidebar">
    <h2>License</h2>
<p><a href="https://www.r-project.org/Licenses/GPL-2">GPL-2</a></p>
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Michael Raboin <br><small class="roles"> Author</small> </li>
<li>R. Mark Sharp, Ph.D. <br><small class="roles"> Author, maintainer </small> </li>
</ul>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Michael Raboin.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
