<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Development Plans • nprcmanager</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Development Plans">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nprcmanager</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.3.21</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Development_Plans.html">Development Plans</a>
    </li>
    <li>
      <a href="../articles/breeding_group_algorithm.html">Breeding Group Algorithm</a>
    </li>
    <li>
      <a href="../articles/breeding_group_formation.html">Breeding Group Formation</a>
    </li>
    <li>
      <a href="../articles/genetic_value_analysis.html">Genetic Value Analysis</a>
    </li>
    <li>
      <a href="../articles/genome_uniqueness_algorithm.html">Algorithm: Genome Uniqueness</a>
    </li>
    <li>
      <a href="../articles/gv_and_bg_desc.html">Genetic Value Analysis and Breeding Group Formation Description</a>
    </li>
    <li>
      <a href="../articles/input.html">Input</a>
    </li>
    <li>
      <a href="../articles/installation.html">Installation</a>
    </li>
    <li>
      <a href="../articles/interactive_use_tutorial.html">Interactive Use of nprcmanager</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/manual.html">Genetic Management Tools Manual</a>
    </li>
    <li>
      <a href="../articles/orip_reporting.html">ORIP Reporting</a>
    </li>
    <li>
      <a href="../articles/pedigree_browser.html">Pedigree Browser</a>
    </li>
    <li>
      <a href="../articles/running_shiny_application.html">Running Shiny Application</a>
    </li>
    <li>
      <a href="../articles/summary_of_major_functions.html">Summary of Major Functions</a>
    </li>
    <li>
      <a href="../articles/summary_statistics.html">Summary Statistics</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Development Plans</h1>
                        <h4 class="author">R. Mark Sharp</h4>
            
            <h4 class="date">Sept. 1, 2017</h4>
      
      
      <div class="hidden name"><code>Development_Plans.Rmd</code></div>

    </div>

    
    
<div id="next-steps" class="section level2">
<h2 class="hasAnchor">
<a href="#next-steps" class="anchor"></a>Next steps</h2>
<ol style="list-style-type: decimal">
<li>Find out what is needed to assign known genotypes to individuals and to incorporate that information into the gene dropping routine. This has been done at the function level in the <em>geneDrop</em> function. The next step is to allow the pedigree submitted via the UI to contain the genetic information. Completed by 20170916.</li>
<li>Find some ways to use genetic marker data in the evaluation of genetic value. This needs to be discussed.</li>
<li><p>Develop a simplistic method of using data directly from LabKey via the labkey-api-r. There is now a configuration file that can be used by the application to get the primate center designation, the LabKey baseUrl, a schemaName, a folderPath, and the queryName. The file is structured so that more than one of each could be placed in the file. However, more design work is needed before this will be of use. Completed initial version 20170907.</p></li>
<li><p>Develop the appropriate UI - Completed 20170916. I got the logging system integrated into the package. I am using the package futile.logger (funny name). Note the checkbox at the bottom of the side panel on the two attached images. When the Debug on checkbox is checked (it is not checked by default), the application writes to a file in the users home directory named nprcmanager.log. Currently I am only logging events occurring the the server.R file, because that is where I tend to have most of my problems show up. I have attached an example file in which turned on logging at the debug level and then read in a couple of pedigree files.</p></li>
</ol>
<p>I added code coverage reports to the automated build system running on Travis-CI.org. Currently we only have 5.68% of the code being tested with the unit tests I created thus far. I do not know the code well enough to know what percentage we will have when I feel comfortable, but I know most of the functions have no tests at all. – Have 86.32% as of 20180429</p>
</div>
<div id="to-do" class="section level2">
<h2 class="hasAnchor">
<a href="#to-do" class="anchor"></a>To do</h2>
<ol style="list-style-type: decimal">
<li>It would be better to catalogue and report the changes made by qcStudbook() so that the user knows how the provided pedigree was changed.</li>
<li>
<p>Amanda Vinson wrote this in an e-mail on 20170901:</p>
<p>I’m very interested in hearing about your idea of adding algorithms related to selecting/maintaining specific alleles in a colony. I think your idea could be related to what I need to work on next, which is an algorithm for assessing the long-term impact on colony genetic diversity of removing specific animals/alleles.<br>
This is a need we have here at the ONPRC due to assigning animals to<br>
research, and I have been tasked with implementing the solution.</p>
</li>
<li><p>Remove individuals that can never breed again from genetic analyses. (From page 29 of <em>PMx User Manual, version 1.0</em>). It is generally recommended to remove living individuals that can never breed or will never be bred again (e.g., post reproductive, sterilized, poor health, hybrid) from the genetic mean kinship calculations, as these individuals can no longer pass on their genetic material to the next generation and their inclusion influences the mean kinship values of other individuals in the population. However, these non-reproductive individuals may occupy cage space and need to be managed in other ways. In addition, their demographic information is valuable. Managers therefore often chose to include these individuals in the demographic analysis but to exclude them from genetic analysis.</p></li>
<li><p>It would be nice to have multiple “selected populations” that could be set up and easily selected for various analyses and plots.</p></li>
<li>
<p>Talk to Amanda about developing a general way to handle animals without two known parents that are not founders. There are the following lines of code in orderReport.R</p>
<blockquote>
<p># ONPRC-born animals with no parentage<br>
i &lt;- (is.na(rpt$origin) &amp; (rpt$totalOffspring == 0) &amp; (rpt$id %in% founders))<br>
noParentage &lt;- rpt[i, ]</p>
</blockquote>
</li>
<li>There are some functions that are defined and not used. They may be needed later.
<ol style="list-style-type: decimal">
<li>findLoops</li>
<li>makeLoop</li>
<li>countLoops</li>
<li>createPedTree</li>
</ol>
</li>
</ol>
<p>I expect they are direct ports from kinship2. I will check.</p>
</div>
<div id="accomplishments-thus-far" class="section level2">
<h2 class="hasAnchor">
<a href="#accomplishments-thus-far" class="anchor"></a>Accomplishments thus far</h2>
<div id="section" class="section level4">
<h4 class="hasAnchor">
<a href="#section" class="anchor"></a>20170911</h4>
<p>I have connected the Travis CI (continuous integration) tool to my github.com/rmsharp/nprcmanager so that when I check in code it automatically tries to build it from scratch using packages from CRAN. Travis CI is now automatically building nprcmanager without errors or warnings. This puts us considerably closer to being able to offer the package via CRAN if we want.</p>
</div>
<div id="section-1" class="section level4">
<h4 class="hasAnchor">
<a href="#section-1" class="anchor"></a>20170915</h4>
<p>Added a genetic uniqueness plot to the Summary Statistics tab.</p>
</div>
<div id="section-2" class="section level4">
<h4 class="hasAnchor">
<a href="#section-2" class="anchor"></a>20170917</h4>
<p>I discovered why the plot of genetic uniqueness scores was not as expected. It was from an analysis where 0 was the threshold setting, which I think we should consider removing as an option. I have set the default to 3, which is what Amando Vinson’s paper indicated that ONPRC typically used.</p>
<p>I got the logging system integrated into the package. I am using the package futile.logger (funny name). Note the checkbox at the bottom of the side panel on the two attached images. When the Debug on checkbox is checked (it is not checked by default), the application writes to a file in the users home directory named <em>nprcmanager.log</em>. Currently I am only logging events occurring the the <em>server.R</em> file, because that is where I tend to have most of my problems show up. I have attached an example file in which turned on logging at the debug level and then read in a couple of pedigree files.</p>
<p>I added code coverage reports to the automated build system running on Travis-CI.org. Currently we only have 5.68% of the code being tested with the unit tests I created thus far. I do not know the code well enough to know what percentage we will have when I feel comfortable, but I know most of the functions have no tests at all.</p>
</div>
<div id="section-3" class="section level4">
<h4 class="hasAnchor">
<a href="#section-3" class="anchor"></a>20170919</h4>
<p>We have successfully installed and used <em>nprcmanager</em> on Microsoft (MS) Windows 7, MS Windows 10, and MacOS 10.12.6 running R 3.4.1.</p>
</div>
<div id="section-4" class="section level4">
<h4 class="hasAnchor">
<a href="#section-4" class="anchor"></a>20170921</h4>
<p>Unit test coverage is about 15%.</p>
</div>
<div id="section-5" class="section level4">
<h4 class="hasAnchor">
<a href="#section-5" class="anchor"></a>20171004</h4>
<p>Unit test coverage is 84.79%.</p>
</div>
<div id="section-6" class="section level4">
<h4 class="hasAnchor">
<a href="#section-6" class="anchor"></a>20180429</h4>
<p>Test coverage is 86.32%.</p>
</div>
</div>
<div id="meeting-notes" class="section level2">
<h2 class="hasAnchor">
<a href="#meeting-notes" class="anchor"></a>Meeting notes</h2>
<div id="meeting" class="section level3">
<h3 class="hasAnchor">
<a href="#meeting" class="anchor"></a>20170323 Meeting</h3>
<p>I met with Jack Kent, Deborah Newman, and Charles Peterson about how to use genetic data in a gene dropping simulation.</p>
<p>From: R. Mark Sharp <a href="mailto:msharp@TxBiomed.org">msharp@TxBiomed.org</a></p>
<p>Subject: Re: Genetic management of colonies</p>
<p>Date: March 23, 2017 at 5:01:07 PM CDT</p>
<p>To: Jack Kent <a href="mailto:jkent@txbiomed.org">jkent@txbiomed.org</a></p>
<p>Cc: Deborah Newman <a href="mailto:dnewman@txbiomedgenetics.org">dnewman@txbiomedgenetics.org</a>, Charles Peterson <a href="mailto:charlesp@txbiomed.org">charlesp@txbiomed.org</a></p>
<p>Jack, Debbie, and Charles,</p>
<p>Thank you for meeting with me. I appreciate your help in clarifying the various stages of the issues associated with providing genetic management guidance via simulation with partially known MHC data.</p>
<ol style="list-style-type: decimal">
<li>
<p>I will look more closely into what will be needed to add genetic information and gene frequency information into the gene dropping routines in the kinship2 package. The initial implementation will use only a single locus, which should be a sufficient for MHC data</p>
<p>A. First step will be to ensure we get expected proportions of genes in children of individuals when we provide genetic data to all founders using a gene frequency based algorithm.</p>
<p>B. Second step will be to ensure we get expected proportions of genes in children when all parental generations have fully known genotypes.</p>
<p>C. Third step will be to ensure we get expected proportions of genes in children when parental generations have partially known genotypes and all other genotypes are uninformative.</p>
<p>D. Fourth step will be to ensure we get expected proportions of genes in children when parental generations have partially known genotypes and remaining parental genes are determined by gene frequency.</p>
</li>
<li><p>We will not worry about getting rid of genes in the pedigrees since colony managers need only not breed animals that carry unwanted alleles.</p></li>
<li><p>I will produce a small number of pedigree drawings using the data Debbie has provided to get feedback on preferences and additional requirements.</p></li>
<li><p>I will not deal with paternity and maternity issues within the pedigrees at this time.</p></li>
</ol>
<p>Thanks again.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#next-steps">Next steps</a></li>
      <li><a href="#to-do">To do</a></li>
      <li><a href="#accomplishments-thus-far">Accomplishments thus far</a></li>
      <li><a href="#meeting-notes">Meeting notes</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Michael Raboin, R. Mark Sharp.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
