<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Algorithm: Genome Uniqueness • nprcmanager</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Algorithm: Genome Uniqueness">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">nprcmanager</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.3.21</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Development_Plans.html">Development Plans</a>
    </li>
    <li>
      <a href="../articles/breeding_group_algorithm.html">Breeding Group Algorithm</a>
    </li>
    <li>
      <a href="../articles/breeding_group_formation.html">Breeding Group Formation</a>
    </li>
    <li>
      <a href="../articles/genetic_value_analysis.html">Genetic Value Analysis</a>
    </li>
    <li>
      <a href="../articles/genome_uniqueness_algorithm.html">Algorithm: Genome Uniqueness</a>
    </li>
    <li>
      <a href="../articles/gv_and_bg_desc.html">Genetic Value Analysis and Breeding Group Formation Description</a>
    </li>
    <li>
      <a href="../articles/input.html">Input</a>
    </li>
    <li>
      <a href="../articles/installation.html">Installation</a>
    </li>
    <li>
      <a href="../articles/interactive_use_tutorial.html">Interactive Use of nprcmanager</a>
    </li>
    <li>
      <a href="../articles/introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/manual.html">Genetic Management Tools Manual</a>
    </li>
    <li>
      <a href="../articles/orip_reporting.html">ORIP Reporting</a>
    </li>
    <li>
      <a href="../articles/pedigree_browser.html">Pedigree Browser</a>
    </li>
    <li>
      <a href="../articles/running_shiny_application.html">Running Shiny Application</a>
    </li>
    <li>
      <a href="../articles/summary_of_major_functions.html">Summary of Major Functions</a>
    </li>
    <li>
      <a href="../articles/summary_statistics.html">Summary Statistics</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Algorithm: Genome Uniqueness</h1>
                        <h4 class="author">Amanda Vinson</h4>
            
            <h4 class="date">10/2/2017</h4>
      
      
      <div class="hidden name"><code>genome_uniqueness_algorithm.Rmd</code></div>

    </div>

    
    
<div id="algorithm-genome-uniqueness" class="section level2">
<h2 class="hasAnchor">
<a href="#algorithm-genome-uniqueness" class="anchor"></a>Algorithm: Genome Uniqueness</h2>
<p>Genome uniqueness is calculated through the use of a gene-drop simulation to estimate how frequently an animal will possess founder alleles not present in other members of the focal population, or present in a specified number or fewer.</p>
<p>The gene-drop simulation used by the web application is a vectorized version and is shown in the figure below. In an un-vectorized version, if 5000 gene-drop simulations are desired for the estimation process, the population had to be iterated over 5000 times. Since each iteration of the gene-drop is independent, the process can be vectorized so that each element of a vector represents 1 iteration of the gene-drop simulation. In the vectorized version, the population is iterated over once, regardless of the number of simulations desired. This drastically reduces the amount of time necessary for the program to run.</p>
<div id="overview" class="section level4">
<h4 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h4>
<p>The basic steps of the gene-drop are:</p>
<ol style="list-style-type: decimal">
<li>Each founder is assigned two unique alleles<br>
</li>
<li>For each subsequent generation:
<ol style="list-style-type: lower-alpha">
<li>Assign genotypes to each member of the generation
<ul>
<li>For each animal, find the genotypes of the parents, and select<br>
one allele from each parent randomly.</li>
</ul>
</li>
</ol>
</li>
</ol>
<p>Once every animal has been assigned a genotype by mendelian inheritance tally the number of unique alleles possessed by each member of the focal population. In the case of this algorithm, we do allow the ‘uniqueness’ threshold to be adjusted so that an allele can be considered unique if it is possessed by N or fewer other members of the focal population.</p>
</div>
<div id="vectorized-gene-drop-details" class="section level4">
<h4 class="hasAnchor">
<a href="#vectorized-gene-drop-details" class="anchor"></a>Vectorized Gene-Drop Details</h4>
<p>The vectorized gene-drop simulation follows the same basic process described above. The difference is that instead of dropping one allele at a time, and repeating the simulation N times, the vectorized version drops N independent alleles one time.</p>
<p>In the vectorized version, each animal has a vector of paternally inherited alleles and a vector of maternally inherited alleles. For each offspring, a random combination of these alleles is produced and dropped down to the offspring by the process below and shown in the following figure:</p>
<ol style="list-style-type: decimal">
<li>To start the simulation, each founder is assigned two unique founding alleles.<br>
N-element vectors are created of these alleles, where N is the desired number of<br>
simulations. In the example below, this founder was assigned the unique founder<br>
alleles 1 &amp; 2 and 5 simulations were desired.<br>
</li>
<li>Each time alleles need to be dropped from parent to offspring, a unique<br>
transmission vector is created representing whether or not an allele<br>
was passed to that offspring. The vector is generated to contain a random combination<br>
of 0’s and 1’s. The animal’s paternally inherited alleles are then multiplied by the<br>
transmission vector, while the maternally inherited alleles are multiplied by the<br>
compliment of the transmission vector.<br>
</li>
<li>To generate the final set of alleles received by the offspring, the maternal and<br>
paternal allele vectors are added together.<br>
</li>
<li>The result is a vector of alleles that this offspring has received from this parent.</li>
</ol>
<p>Once allele vectors have been generated for every animal in the pedigree, the focal population can be subset out. Within this population of allele vectors, unique alleles can be determined:</p>
<p>For each position on the allele vectors (1:N) - Gather each animal’s two alleles - If the number of other animals possessing that allele is equal to, or below the threshold, score the allele as unique (1) - Otherwise, score the allele as non-unique (0)</p>
<p>Once every position on each animal’s two allele vector’s has been scored, sum all of the scores for an animal and divide by the total number of alleles being considered (2 * number of simulations).</p>
<div class="figure">
<img src="../inst/application/www/GeneDrop.png" alt="Generation of a vector of five gametes from one parent. Showing how the transmission vectors (row 2) determine which alleles are passed from the parental alleles or haplotypes (row 1) to form complementary vectors (row 3) that are combined by adding corresponding elements to form the final vector of transmitted alleles (row 4)."><p class="caption">Generation of a vector of five gametes from one parent. Showing how the transmission vectors (row 2) determine which alleles are passed from the parental alleles or haplotypes (row 1) to form complementary vectors (row 3) that are combined by adding corresponding elements to form the final vector of transmitted alleles (row 4).</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#algorithm-genome-uniqueness">Algorithm: Genome Uniqueness</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Michael Raboin, R. Mark Sharp.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
