---
title: "Shiny Application for Genetic Management of Colonies"
author: "R. Mark Sharp"
output:
  html_document:
    df_print: paged
vignette: |
  %\VignetteEngine{knitr::knitr} 
  %\VignetteIndexEntry{Genetic Management Tools -- Shiny App} 
  %\usepackage[UTF-8]{inputenc}
---

```{r setup, include=FALSE}
library(png)
library(grid)
library(stringi)
library(nprcmanager)
knitr::opts_chunk$set(eval = FALSE, echo=TRUE, results = "markup", cache = TRUE)
start_time <- proc.time()
```


## Introduction
This tutorial demonstrates the major functions used 
within the Shiny application provided by the __nprcmanager__ package.
This is a brief tutorial that illustrates a typical workflow and does
not explore all possible workflows.

Please provide any comments, questions, or bug reports through the GitHub
issue tracker at 
[https://github.com/rmsharp/nprcmanager/issues](
https://github.com/rmsharp/nprcmanager/issues).

## Installation and Help

To get the most recent version you can install **nprcmanager** from
GitHub with the following code.

```{r gh-installation}
install.packages("devtools")
devtools::install_github("rmsharp/nprcmanager")

```

All missing dependencies should be automatically installed.

You can start the Shiny application from the __R__ console with 

```{r start, echo = TRUE, include=TRUE}
library(nprcmanager)
#runManager()
getwd()
```

This will result in the opening screen where you tell the application how to 
find the pedigree you will be using.
Most of the screen is filled with information about formatting the pedigree 
file.


*********

![](../../inst/extdata/shiny_app_tutorial/images/opening_screen_top.png)

*********


Scrolling down to the middle of the opening screen exposes a table that 
describes a pedigree file and further instructions.

*********

![](../../inst/extdata/shiny_app_tutorial/images/opening_screen_middle.png)

*********


Scrolling down to the bottom of the opening screen exposes more pedigree file 
instructions, a table that describes a genotype file and instructions regarding
use of a genotype file. This tutorial will not include instructions on using a 
genotype file.

*********


![](../../inst/extdata/shiny_app_tutorial/images/opening_screen_bottom.png)


*********

In this introductory tutorial, we will use an Excel file containing 
a hypothetical pedigree of macques.
We will work with the gray box on the left at the top of the screen.


*********

![](../../inst/extdata/shiny_app_tutorial/images/opening_screen_top_red_oval.png)

*********



A Microsoft Excel workbook with a single worksheet is the default file 
type; though comma (.csv), semi-colon (.txt), and tab (.txt) separated value 
files are all acceptable formats.

The _Example_Pedigree.xlsx_ file we are using is from a CSV file created as 
shown below and then saved in an Excel format.
```{r make-example-file}
makeExamplePedigreeFile()
```

Select the __Browse__ button and select the pedigree file from your file system.


*********
```{r example-pedigree, eval = TRUE, fig.width = 2.5, fig.height = 3.5, echo = FALSE}
img <- readPNG("../../inst/extdata/shiny_app_tutorial/images/input_example_pedigree.xlsx.png")
grid.raster(img)
```

*********


It is important to make sure the minimum parent age is low enough for the 
animals in your pedigree. For our example pedigree, we are changing it from
4 years to 2 years of age since these macques may reproduce as early as two 
years of age.
This is shown below in three progressive images with the center image 
demonstrating how the hovertext provides an explaination of how this value is 
used.


*********

```{r example-pedigree-minParentAgeSequence, eval = TRUE, fig.width = 10, fig.height = 5, echo = FALSE}
img <- readPNG("../../inst/extdata/shiny_app_tutorial/images/input_minParentAgeSequence.png")
grid.raster(img)
```


*********



## Reading in a Pedigree and Testing for Errors

Selected __Read and Check Pedigree__ will read in the file and test to see
if the pedigree file has all of the columns needed and the pedigree is 
internally consistent.

*********

```{r read-and-check-pedigree, eval = TRUE, fig.width = 2.5, fig.height = 3.5, echo = FALSE}
img <- readPNG("../../inst/extdata/shiny_app_tutorial/images/read_and_check_pedigree.png")
grid.raster(img)
```


*********



Several error types, shown below, are detected the application.

```{r make-errorList-definition-tbl, echo = FALSE, eval=TRUE}
errorTypes <- names(getEmptyErrorLst())
errorDescriptions <- c(
  "Database connection failed: configuration or permissions are invalid",
  "Columns that must be within the pedigree file are missing.",
  "Values, which are supposed to be dates, cannot be interpreted as a date.",
  "Parents were too young on the date of birth of to have been the parent.",
  "Individuals listed as female or hemaphroditic and as a sire.",
  "Individuals are listed as male and as a dam.", 
  "Individuals who are listed as both a sire and a dam.",
  "IDs listed more than once.",
  stri_c("Columns that have been changed to conform to internal naming ",
  "conventions and what they were changed to.")
)
errorTbl <- data.frame(`Error` = errorTypes, Definition = errorDescriptions, 
                       stringsAsFactors = FALSE)

```
```{r print-error-definition-tbl, eval = TRUE, echo=FALSE, include = TRUE, results='markup'}
knitr::kable(errorTbl) 

```

*********


The __Pedigree Browser__ tab defaults to displaying 10 rows of the pedigree
at a time, but you can choose to display 10, 25, 50, or 100 rows.
You can choose to display UNKNOWN IDs in the rows displayed.



![](../../inst/extdata/shiny_app_tutorial/images/pedigree_browser_100_rows_display_unknown_ids.png)


*********

I have place red lines under the UNKNOWN IDs in the partial pedigree list below
for clarity. These IDs have no meaning other than they all begin with the letter
_U_ and are following with a left zero-filled integer of four places.


*********

![](../../inst/extdata/shiny_app_tutorial/images/pedigree_browser_unknown_displayed.png)


*********

In this example pedigree, when you deselect the __Display Unknown IDs__
checkbox.
The number of rows reduces from 3,694 to 2,322, because there were 1,372
UNKNOWN animals generated when constructing the pedigree to provide sire
and dam placeholders for all animals.

*********


![](../../inst/extdata/shiny_app_tutorial/images/pedigree_browser_no_unknown_displayed.png)

*********


The __Pedigree Browser__ tab displays the full pedigree by default but allows
you to select a subset of the pedigree by entering a list of 
animals of interest (_focal animals_).

*********


![](../../inst/extdata/shiny_app_tutorial/images/pedigree_browser_focal_animal_text_box.png)

*********


You can enter in the animal IDs by typing them into the text box directly as
shown below.

*********
```{r pedigree-browser-5-focal-animals-small, eval = TRUE, fig.width = 3, fig.height = 1, echo = FALSE}
img <- readPNG("../../inst/extdata/shiny_app_tutorial/images/pedigree_browser_5_focal_animals_small.png")
grid.raster(img)
```


*********

After entering your list of focal animals, you can select to trim the 
pedigree so that it will only include relatives of the focal animals you have 
selected.


*********


![](../../inst/extdata/shiny_app_tutorial/images/pedigree_browser_select_trim_for_focal_animals.png)

*********


Also, you can import a list of focal animals by selecting the __Browse__ button
under __Choose CSV file with focal animals__. 
This file can be constructed by creating a simple text file with commas between
animal IDs or by placing individual animal IDs on separate lines.


*********

![](../../inst/extdata/shiny_app_tutorial/images/pedigree_browser_selection_large_focal_group.png)

*********

A pedigree trimmed based on focal animals will have only the relatives of those
animals remaining.
In this instance there are only a total of 85 focal animals and their relatives.
*********



![](../../inst/extdata/shiny_app_tutorial/images/pedigree_browser_trimmed_for_focal_animals.png)

*********


The __Pedigree Age Plot__ tab displays a standard pyramid plot for the pedigree
as selected in the __Pedigree Browser__ tab.
This is showing 332 living animals from the entire example pedigree.
![](../../inst/extdata/shiny_app_tutorial/images/age_plot.png)

*********


Selecting the __Genetic Value Analysis__ tab and the __Begin Analysis__ button will
begin the gene dropping process, which you can monitor with the progress meter
in the lower right corner of the display. 

We are not aware of a systematic study of pedigree structure with this 
algorithm and have not performed extensive studies with pedigrees of various 
structures, but 1000 iterations has seemed to provide reproducible results. 
 

*********


![](../../inst/extdata/shiny_app_tutorial/images/genetic_value_analysis_calculating.png)

*********


As soon as the calculations are completed, a table showing
the results of the analysis is displayed in 10 record increments. 
The calculations for 1000 iterations of the gene dropping algorithm took 1 
minute 38 seconds with the example pedigree of 3,691
animals using a MacBook Pro (Mid 2014), 2.8 GHz Intel Core i7 with 16 GB of
1600 MHz DDR3 memory.

Again you can select how many rows to display at once by changing the values in the 
__Show entries__ selection tool.

*********

![](../../inst/extdata/shiny_app_tutorial/images/genetic_value_analysis_first_high_value.png)

*********

Searching down the table of results in the __Value Designation__ column you can 
see starting at row 2777 the values change from _High Value_ to _Low Value_.


*********

![](../../inst/extdata/shiny_app_tutorial/images/genetic_value_analysis_high_and_low_value.png)


*********
The value of _Undetermined_ in the __Value Designation__ column means the 
animal did not have parentage information.

*********

![](../../inst/extdata/shiny_app_tutorial/images/genetic_value_analysis_undetermined_value.png)

*********
The __Summary Statisitics and Plots__ tab used results from the 
__Genetic Value Analysis__ tab.

*********

![](../../inst/extdata/shiny_app_tutorial/images/summary_statistics_first_view.png)
*********

The six plots provide histograms and boxplots for the kinship coefficients,
the Z-scores of the kinship coefficients, and the genome uniqueness scores.


*********

![](../../inst/extdata/shiny_app_tutorial/images/summary_statistics_all_plots.png)

*********

Selecting the __Breeding Group Formation__ tab brings forward the screen shown
below. 
In this screen you can form breeding groups using one of three ways based on
your source of animals selected under __Choose one group formation workflow:__.

Further you must specify how you want to construct the breeding groups with 
regard to the groups' sex ratios. 
The third of the three options (_User specified sex ratio of breeders_) causes
the appearance of the field where you can fill in the sex ratio (F/M) that 
you want to have in the formed breeding groups.
The sex ratio algorithm will form a group as nearly to the selected ratio as 
possible given the size of the group. 
Limits in the availability of either sex will restrict the size of the groups 
formed.

*********

![](../../inst/extdata/shiny_app_tutorial/images/breeding_group_1.png)


*********


*********


![](../../inst/extdata/shiny_app_tutorial/images/breeding_group_6_seed_groups.png)

*********


*********



![](../../inst/extdata/shiny_app_tutorial/images/breeding_group_6_seed_grps_grp_6_kinship.png)


*********


*********


![](../../inst/extdata/shiny_app_tutorial/images/breeding_group_1.png)

*********



















```

